---
output: html_document
editor_options: 
  chunk_output_type: console
---
# A complete bioinformatics workflow in R

<!-- TODO for Dr. B: create function to simplify the prepwork -->

**By**: Nathan L. Brouwer

# "Worked example: Building a phylogeny in R" 

## Introduction

Phylogenies play an important role in computational biology and bioinformatics.  Phylogenetics itself is an obligatly computational field that only began rapid growth when computational power allowed the many algorithms it relies on to be done rapidly.  Phylogeneies of species, genes and proteins are used to address many biological issues, including

* Patterns of protein evolution
* Origin and evolution of phenotypic traits
* Origin and progression of epidemics
* Origin of evolution of diseases (eg, zooenoses)
* Prediction of protein function from its sequence
* ... and many more

The actual building of a phylogeny is a computationally intensive task; moreover, there are many bioinformatics and computational tasks the precede the construction of a phylogeny:

* genome sequencing and assembly
* computational gene prediction and annotation
* database searching and results screening
* pairwise sequence alignment
* data organization and cleaning
* multiple sequence alignment
* evaluation and validation of alignment accuracy

Once all of these steps have been carried out, the building of a phylogeny involves

* picking a model of sequence evolution or other description of evolution
* picking a statistical approach to tree construction
* evaluating uncertainty in the final tree

In this chapter we will work through many of these steps.  In most cases we will pick the easiest or fastest option; in later chapters we will unpack the various options.  This chapter is written as an interactive R sessions.  You can follow along by opening the .Rmd file of the chapter or typing the appropriate commands into your own script.  I assume that all the necessary packages have been installed and they only need to be loaded into R using the *library()* command.


This lesson walks you through and entire workflow for a bioinformatics, including

<!-- check all of this -->

1. obtaining FASTA sequences
1. cleaning sequences
1. creating alignments
1. creating distance a distance matrix
1. building a phylogenetic tree

We'll examine the Shroom family of genes, which produces Shroom proteins essential for tissue formation in many multicellular eukaryotes, including neural tube formation in vertebrates.  We'll examine shroom in several very different organism, including humans, mice and sea urchins.  There is more than one type of shroom in vertebrates, and we'll also look at two different Shroom genes: shroom 1 and shroom 2.

This lesson draws on skills from previous sections of the book, but is written to act as a independent summary of these activities.  There is therefore review of key aspects of R and bioinformatics throughout it.


### Vocab 

## Software Preliminaires


* argument
* function
* list
* named list
* vector
* named vector
* for() loop
* R console



### R functions

* library()
* round()
* plot()
* mtext()
* nchar()
* rentrez::entrez_fetch()
* compbio4all::entrez_fetch_list()
* compbio4all::print_msa() (Coghlan 2011)
* Biostrings::AAStringSet()
* msa::msa()
* msa::msaConvert()
* msa::msaPrettyPrint()
* seqinr::dist.alignment()
* ape::nj()

A few things need to be done to get started with our R session.

### Download necessary packages

Many *R* sessions begin by downloading necessary software packages to augment *R's* functionality.

If you don't have them already, you'll need the following packages from CRAN:

1. `ape`
1. `seqinr`
1. `rentrez`
1. `devtools`

The CRAN packages can be loaded with `install.packages()`.  

You'll also need these packages from Bioconductor:

1. `msa`
1. `Biostrings`

For installing packages from Bioconductor, see the chapter at the beginning of this book on this process.

Finally, you'll need this package from GitHub

1. `compbio4all`
1. `ggmsa`


To install packages from GitHub you can use the code `devtools::install_github("brouwern/combio4all")` and `devtools::install_github("YuLab-SMU/ggmsa")`

### Load packages into memory

We now need to load up all our bioinformatics and phylogenetics software into R.  This is done with the `library()` command.  

To run this code just clock on the sideways green triangle all the way to the right of the code.

NOTE: You'll likely see some red code appear on your screen.  No worries, totally normal!

```{r, message= F, warning=F}
# github packages
library(compbio4all)

# CRAN packages
library(rentrez)
library(seqinr)
library(ape)

# Bioconductor packages
library(msa)
library(Biostrings)
```


## Downloading macro-molecular sequences

We're going to explore some sequences.  First we need to download them.  To do this we'll use a function, `entrez_fretch()`, which accesses the [**Entrez**](ncbi.nlm.nih.gov/search/) system of database (ncbi.nlm.nih.gov/search/).  This function is from the `rentrez` package, which stands for "R-Entrez."

We need to tell `entrez_fetch()` several things

1. `db = ...` the type of entrez database.  
1. `id = ...` the **accession** (ID) number of the sequence
1. `rettype = ...` file type what we want the function to return.  

Formally, these things are called **arguments** by *R*.


We'll use these settings:

1. `db = "protein"` to access the Entrez database of protein sequences  
1. `rettype = "fasta"`, which is a standard file format for nucleic acid and protein sequences

We'll set `id = ...` to sequences whose **accession numbers** are:

1. NP_065910: Human shroom 3
1. AAF13269:  Mouse shroom 3a
1. CAA58534:  Human shroom 2
1. XP_783573:  Sea urchin shroom

There are two highly conserved regions of shroom 3
1. ASD 1: aa 884 to aa 1062 in  hShroom3
1. ASD 2: aa 1671 to aa 1955 in hShroom3


Normally we'd have to download these sequences by hand through pointing and clicking on GeneBank records on the NCBI website.  In *R* we can do it automatically; this might take a second.

All the code needed is this:
```{r}
# Human shroom 3 (H. sapiens)
hShroom3 <- entrez_fetch(db = "protein", 
                          id = "NP_065910", 
                          rettype = "fasta")
```

The output is in FASTA format; we'll use the `cat()` to do a little formatting for us:
```{r}
cat(hShroom3)
```

Note the initial `>`, then the header line of `NP_065910.3 protein Shroom3 [Homo sapiens]`.  After that is the amino acid sequence.  The underlying data also includes the **newline character** `\n` to designate where each line of amino acids stops.


We can get the rest of the data by just changing the `id = ...` argument:
```{r}
# Mouse shroom 3a (M. musculus)
mShroom3a <- entrez_fetch(db = "protein", 
                          id = "AAF13269", 
                          rettype = "fasta")

# Human shroom 2 (H. sapiens)
hShroom2 <- entrez_fetch(db = "protein", 
                          id = "CAA58534", 
                          rettype = "fasta")


# Sea-urchin shroom
sShroom <- entrez_fetch(db = "protein", 
                          id = "XP_783573", 
                          rettype = "fasta")
```


I'm going to check about how long each of these sequences is - each should have an at least slightly different length. If any are identical, I might have repeated an accession name or re-used an object name. The function `nchar()` counts of the number of characters in an *R* object.
```{r}
nchar(hShroom3)
nchar(mShroom3a)
nchar(sShroom)
nchar(hShroom2)
```



## Prepping macromolecular sequences

>"90% of data analysis is data cleaning" (-Just about every data analyst and data scientist on twitter)

We have our sequences, but the current format isn't directly usable for us yet because there are several things that aren't sequence information

1. metadata (the header)
1. page formatting information (the newline character)

We can remove this non-sequence information using a function I wrote called `fasta_cleaner()`, which is in the `compbio4all` package. The function uses **regular expressions** to remove the info we don't need.

If you had trouble downloading the fasta_cleaner() function you can add it to your R session by running this code:

```{r}
fasta_cleaner <- function(fasta_object, parse = TRUE){

  fasta_object <- sub("^(>)(.*?)(\\n)(.*)(\\n\\n)","\\4",fasta_object)
  fasta_object <- gsub("\n", "", fasta_object)

  if(parse == TRUE){
    fasta_object <- stringr::str_split(fasta_object,
                                       pattern = "",
                                       simplify = FALSE)
  }

  return(fasta_object[[1]])
}

```


**ASIDE**: If we run the name of the command with out any quotation marks we can see the code:
```{r}
fasta_cleaner
```
**End ASIDE**


Now use the function to clean our sequences; we won't worry about what `pare = ...` is for.
```{r}
hShroom3  <- fasta_cleaner(hShroom3,  parse = F)
mShroom3a <- fasta_cleaner(mShroom3a, parse = F)
hShroom2  <- fasta_cleaner(hShroom2,  parse = F)
sShroom   <- fasta_cleaner(sShroom,   parse = F)
```


Now let's take a peek at what our sequences look like:
```{r}
hShroom3
```


## Aligning sequences

We can do a [**global alignment**](https://tinyurl.com/y4du73zq) of one sequence against another using the `pairwiseAlignment()` function from the **Bioconductor** package `Biostrings` (note that capital "B" in `Biostrings`; most *R* package names are all lower case, but not this one).

Let's align human versus mouse shroom:
```{r}
align.h3.vs.m3a <- Biostrings::pairwiseAlignment(
                  hShroom3,
                  mShroom3a)
```

We can peek at the alignment
```{r}
align.h3.vs.m3a
```

The **score** tells us how closely they are aligned; higher scores mean the sequences are more similar.  Its hard to interpret the number on its own so we can get the **percent sequence identity (PID)** using the `pid()` function.

```{r}
Biostrings::pid(align.h3.vs.m3a)
```

So, *shroom3* from humans and *shroom3* from mice are ~71% similar (at least using this particular method of alignment, and there are many ways to do this!)

What about human shroom 3 and sea-urchin shroom?

```{r}
align.h3.vs.h2 <- Biostrings::pairwiseAlignment(
                  hShroom3,
                  hShroom2)
```


First check out the score using `score()`, which accesses it directly without all the other information.
```{r}
score(align.h3.vs.h2)
```

Now the percent sequence alignment with `pid()`:
```{r}
Biostrings::pid(align.h3.vs.h2)
```

So Human shroom 3 and Mouse shroom 3 are 71% identical, but Human shroom 3 and human shroom 2 are only 34% similar?  How does it work out evolutionary that a human and mouse gene are more similar than a human and a human gene?  What are the evolutionary relationships among these genes within the shroom gene family?


## The shroom family of genes

I've copied a table from a published paper which has accession numbers for 15 different Shroom genes.

```{r}
shroom_table <- c("CAA78718" , "X. laevis Apx" ,         "xShroom1",
                  "NP_597713" , "H. sapiens APXL2" ,     "hShroom1",
                  "CAA58534" , "H. sapiens APXL",        "hShroom2",
                  "ABD19518" , "M. musculus Apxl" ,      "mShroom2",
                  "AAF13269" , "M. musculus ShroomL" ,   "mShroom3a",
                  "AAF13270" , "M. musculus ShroomS" ,   "mShroom3b",
                  "NP_065910", "H. sapiens Shroom" ,     "hShroom3",
                  "ABD59319" , "X. laevis Shroom-like",  "xShroom3",
                  "NP_065768", "H. sapiens KIAA1202" ,   "hShroom4a",
                  "AAK95579" , "H. sapiens SHAP-A" ,     "hShroom4b",
                 #"DQ435686" , "M. musculus KIAA1202" ,  "mShroom4",
                  "ABA81834" , "D. melanogaster Shroom", "dmShroom",
                  "EAA12598" , "A. gambiae Shroom",      "agShroom",
                  "XP_392427" , "A. mellifera Shroom" ,  "amShroom",
                  "XP_783573" , "S. purpuratus Shroom" , "spShroom") #sea urchin
```


I'll do a bit of formatting; you can ignore these details if you want
```{r}
# convert to matrix
shroom_table_matrix <- matrix(shroom_table,
                                  byrow = T,
                                  nrow = 14)
# convert to dataframe
shroom_table <- data.frame(shroom_table_matrix, 
                     stringsAsFactors = F)

# name columns
names(shroom_table) <- c("accession", "name.orig","name.new")

# Create simplified species names
shroom_table$spp <- "Homo"
shroom_table$spp[grep("laevis",shroom_table$name.orig)] <- "Xenopus"
shroom_table$spp[grep("musculus",shroom_table$name.orig)] <- "Mus"
shroom_table$spp[grep("melanogaster",shroom_table$name.orig)] <- "Drosophila"
shroom_table$spp[grep("gambiae",shroom_table$name.orig)] <- "mosquito"
shroom_table$spp[grep("mellifera",shroom_table$name.orig)] <- "bee"
shroom_table$spp[grep("purpuratus",shroom_table$name.orig)] <- "sea urchin"

```


Take a look:
```{r}
shroom_table
```

## Downloading multiple sequences 

Instead of getting one sequence at a time we can download several by accessing the "accession" column from the table

```{r}
shroom_table$accession
```


We can give this whole set of accessions to `entrez_fetch()`:
```{r}
shrooms <- entrez_fetch(db = "protein", 
                          id = shroom_table$accession, 
                          rettype = "fasta")
```

We can look at what we got here with `cat()`  (I won't display this because it is very long!)
```{r, eval = F}
cat(shrooms)
```

The current format of these data is a single, long set of data.  This is a standard way to store, share and transmit FASTA files, but in *R* we'll need a slightly different format. 

We'll download all of the sequences again, this time using a function from `compbio4all` called `entrez_fetch_list()` which is a **wrapper** function I wrote to put the output of `entrez_fetch()` into an *R* data format called a **list**.

```{r}
shrooms_list <- entrez_fetch_list(db = "protein", 
                          id = shroom_table$accession, 
                          rettype = "fasta")
```

Now we have a list which as 14 **elements**, one for each sequence in our table.

```{r}
length(shrooms_list)
```

We now need to clean up each one fo these sequences.  We can do that using a simple `for()` loop:

```{r}
for(i in 1:length(shrooms_list)){
  shrooms_list[[i]] <- fasta_cleaner(shrooms_list[[i]], parse = F)
}
```

Second to last  step: we need to take each one of our sequences from our list and put it into a **vector**, in particular a **named vector**

```{r}
# make a vector to store output
shrooms_vector <- rep(NA, length(shrooms_list))

# run the loop
for(i in 1:length(shrooms_vector)){
  shrooms_vector[i] <- shrooms_list[[i]]
}

# name the vector
names(shrooms_vector) <- names(shrooms_list)
```

Now the final step: we need to convert our named vector to a **string set** using `Biostrings::AAStringSet()`.  Note the `_ss` tag at the end of the object we're assigning the output to, which designates this as a string set.

```{r}
shrooms_vector_ss <- Biostrings::AAStringSet(shrooms_vector)
```


## Multiple sequence alignment

We must **align** all of the sequences we downloaded and use that **alignment** to build a **phylogenetic tree**.  This will tell us how the different genes, both within and between species, are likely to be related.  




### Building an Multiple Sequence Alignment (MSA)

We'll use the software `msa,` which implements the **ClustalW** multiple sequence alignment algorithm.  Normally we'd have to download the ClustalW program and either point-and-click our way through it or use the **command line***, but these folks wrote up the algorithm in R  so we can do this with a line of R code.  This will take a second or two.
```{r}
shrooms_align <- msa(shrooms_vector_ss,
                     method = "ClustalW")
```

```{r}
str(shrooms_align)
```


### Viewing an MSA

Once we build an MSA we need to visualize it.  

#### Viewing an MSA in R

We can look at the output from `msa()`, but its not very helpful
```{r}
shrooms_align
```

A function called `print_msa()` (Coghlan 2011) which I've put into`combio4all` can give us more informative output by printing out the actual alignment into the R console.  

To use `print_msa()` We need to make a few minor tweaks though first.  These are behind the scenes changes so don't worry about the details right now.  We'll change the name to `shrooms_align_seqinr` to indicate that one of our changes is putting this into a format defined by the bioinformatics package `seqinr`.
```{r}
class(shrooms_align) <- "AAMultipleAlignment"
shrooms_align_seqinr <- msaConvert(shrooms_align, type = "seqinr::alignment")
```


I won't display the output from `shrooms_align_seqinr` because its very long; we have 14 shroom genes, and shroom happens to be a rather long gene.
<!-- TODO: set up for displaying only part of the alignment, subsetting it etc. -->
```{r, eval = F}
print_msa(alignment = shrooms_align_seqinr, 
          chunksize = 60)
```

#### Displaying an MSA as an R plot


I'm going to just show about 100 amino acids near the end of the alignment, where there is the most overlap across all of the sequences.  This is set with the `start = ...` and `end = ...` arguments.  Note that we're using the `shrooms_align` object.

```{r}
ggmsa::ggmsa(shrooms_align,   # shrooms_align, NOT shrooms_align_seqinr
      start = 2000, 
      end = 2100) 
```



#### Saving an MSA as PDF

We can take a look at the alignment in PDF format if we want.  I this case I'm going to just show about 100 amino acids near the end of the alignment, where there is the most overlap across all of the sequences.  This is set with the `y = c(...)` argument.

```{r, eval = F}
msaPrettyPrint(shrooms_align,             # alignment
               file = "shroom_msa.pdf",   # file name
               y=c(2000, 2100),           # range
               askForOverwrite=FALSE)
```

You can see where R is saving things by running `getwd()`
```{r}
getwd()
```

On a Mac you can usually find the file by searching in Finder for the file name, which I set to be "shroom_msa.pdf" using the `file = ...` argument above.



## Genetic distance.

Next  need to first get an estimate of how similar each sequences is.  The more amino acids that are identical to each other, the more similar.  

Instead of similarity, we usually work in terms of *difference* or **genetic distance** (a.k.a. **evolutionary distance**).  This is done with the `dist.alignment()` function.
```{r}
shrooms_dist <- seqinr::dist.alignment(shrooms_align_seqinr, 
                                       matrix = "identity")


```

We've made a matrix using `dist.alignment()`; let's round it off so its easier to look at using the `round()` function.
```{r}
shrooms_dist_rounded <- round(shrooms_dist,
                              digits = 3)
```

Now let's look at it
```{r}
shrooms_dist_rounded
```

## Phylognetic trees (finally!)

We got our sequence, built multiple sequence alignment, and calculated the genetic distance between sequences.  Now we are - finally - ready to build a phylogenetic tree.

First, we let R figure out the structure of the tree.  There are **MANY** ways to build phylogenetic trees.  We'll use a common one used for exploring sequences called **neighbor joining** algorithm via the function `nj()`.  Neighbor joining uses genetic distances to cluster sequences into **clades**.  

```{r}
tree <- nj(shrooms_dist)
```

### Plotting phylogenetic trees 

Now we'll make a quick plot of our tree using `plot()` (and add a little label using a function called `mtext()`).
```{r}
# plot tree
plot.phylo (tree, main="Phylogenetic Tree", 
            type = "unrooted", 
            use.edge.length = F)

# add label
mtext(text = "Shroom family gene tree - unrooted, no branch lengths")
```

This is an **unrooted tree*.  For the sake of plotting we've also ignored the evolutionary distance between the sequences.

To make a rooted tree we remove `type = "unrooted`.

```{r}
# plot tree
plot.phylo (tree, main="Phylogenetic Tree", 
            use.edge.length = F)

# add label
mtext(text = "Shroom family gene tree - rooted, no branch lenths")
```


We can include information about branch length by setting `use.edge.length = ...` to `T`. 

```{r}
# plot tree
plot.phylo (tree, main="Phylogenetic Tree", 
            use.edge.length = T)

# add label
mtext(text = "Shroom family gene tree - rooted, with branch lenths")
```

Some of the branches are now very short, but most are very long, indicating that these genes have been evolving independently for many millions of years.

Let's make a fancier plot.  Don't worry about all the steps; I've added some more code to add some annotations on the right-hand side to help us see what's going on.

<!-- TODO: how does this render? -->
```{r fig.width=6}
plot(tree, main="Phylogenetic Tree")
mtext(text = "Shroom family gene tree")

x <- 0.551
x2 <- 0.6

# label Shrm 3
segments(x0 = x, y0 = 1, 
         x1 = x, y1 = 4,
         lwd=2)
text(x = x*1.01, y = 2.5, "Shrm 3",adj = 0)

segments(x0 = x, y0 = 5, 
         x1 = x, y1 = 6,
         lwd=2)
text(x = x*1.01, y = 5.5, "Shrm 2",adj = 0)

segments(x0 = x, y0 = 7, 
         x1 = x, y1 = 9,
         lwd=2)
text(x = x*1.01, y = 8, "Shrm 1",adj = 0)

segments(x0 = x, y0 = 10, 
         x1 = x, y1 = 13,
         lwd=2)
text(x = x*1.01, y = 12, "Shrm ?",adj = 0)


segments(x0 = x, y0 = 14, 
         x1 = x, y1 = 15,
         lwd=2)
text(x = x*1.01, y = 14.5, "Shrm 4",adj = 0)


segments(x0 = x2, y0 = 1, 
         x1 = x2, y1 = 6,
         lwd=2)

segments(x0 = x2, y0 = 7, 
         x1 = x2, y1 = 9,
         lwd=2)

segments(x0 = x2, y0 = 10, 
         x1 = x2, y1 = 15,
         lwd=2)

```


<!-- Assignment - redo with DNA; how does PID compare -->
<!-- compare pairwise PID to MSA similarity -->


