<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 18 “Worked example: Building a phylogeny in R” | A Little Book of R for Bioinformatics 2.0</title>
  <meta name="description" content="An entry-level text for bioinformatics and computational biology" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 18 “Worked example: Building a phylogeny in R” | A Little Book of R for Bioinformatics 2.0" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="An entry-level text for bioinformatics and computational biology" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 18 “Worked example: Building a phylogeny in R” | A Little Book of R for Bioinformatics 2.0" />
  
  <meta name="twitter:description" content="An entry-level text for bioinformatics and computational biology" />
  

<meta name="author" content="Avril Coghlan, with contributions by Nathan L. Brouwer" />


<meta name="date" content="2022-05-10" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="a-complete-bioinformatics-workflow-in-r.html"/>
<link rel="next" href="simple-for-loop-example.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Little Book of R for Bioinformatics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface to version 2.0</a></li>
<li class="chapter" data-level="1" data-path="downloadR.html"><a href="downloadR.html"><i class="fa fa-check"></i><b>1</b> Downloading R</a>
<ul>
<li class="chapter" data-level="1.1" data-path="downloadR.html"><a href="downloadR.html#preface"><i class="fa fa-check"></i><b>1.1</b> Preface</a></li>
<li class="chapter" data-level="1.2" data-path="downloadR.html"><a href="downloadR.html#introduction-to-r"><i class="fa fa-check"></i><b>1.2</b> Introduction to R</a></li>
<li class="chapter" data-level="1.3" data-path="downloadR.html"><a href="downloadR.html#installing-r"><i class="fa fa-check"></i><b>1.3</b> Installing R</a>
<ul>
<li class="chapter" data-level="1.3.1" data-path="downloadR.html"><a href="downloadR.html#installing-r-on-a-windows-pc"><i class="fa fa-check"></i><b>1.3.1</b> Installing <em>R</em> on a Windows PC</a></li>
<li class="chapter" data-level="1.3.2" data-path="downloadR.html"><a href="downloadR.html#how-to-install-r-on-non-windows-computers-eg.-macintosh-or-linux-computers"><i class="fa fa-check"></i><b>1.3.2</b> How to install <em>R</em> on non-Windows computers (eg. Macintosh or Linux computers)</a></li>
</ul></li>
<li class="chapter" data-level="1.4" data-path="downloadR.html"><a href="downloadR.html#starting-r"><i class="fa fa-check"></i><b>1.4</b> Starting <em>R</em></a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="installing-the-rstudio-ide-installrstudio.html"><a href="installing-the-rstudio-ide-installrstudio.html"><i class="fa fa-check"></i><b>2</b> Installing the RStudio IDE {$installRStudio}</a>
<ul>
<li class="chapter" data-level="2.1" data-path="installing-the-rstudio-ide-installrstudio.html"><a href="installing-the-rstudio-ide-installrstudio.html#getting-to-know-rstudio"><i class="fa fa-check"></i><b>2.1</b> Getting to know RStudio</a></li>
<li class="chapter" data-level="2.2" data-path="installing-the-rstudio-ide-installrstudio.html"><a href="installing-the-rstudio-ide-installrstudio.html#rstudio-versus-rstudio-cloud"><i class="fa fa-check"></i><b>2.2</b> RStudio versus RStudio Cloud</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="installing-packages.html"><a href="installing-packages.html"><i class="fa fa-check"></i><b>3</b> Installing <em>R</em> packages</a>
<ul>
<li class="chapter" data-level="3.1" data-path="installing-packages.html"><a href="installing-packages.html#downloading-packages-with-the-rstudio-ide"><i class="fa fa-check"></i><b>3.1</b> Downloading packages with the RStudio IDE</a></li>
<li class="chapter" data-level="3.2" data-path="installing-packages.html"><a href="installing-packages.html#downloading-packages-with-the-function-install.packages"><i class="fa fa-check"></i><b>3.2</b> Downloading packages with the function <code>install.packages()</code></a></li>
<li class="chapter" data-level="3.3" data-path="installing-packages.html"><a href="installing-packages.html#using-packages-after-they-are-downloaded"><i class="fa fa-check"></i><b>3.3</b> Using packages after they are downloaded</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html"><i class="fa fa-check"></i><b>4</b> Installing Bioconductor</a>
<ul>
<li class="chapter" data-level="4.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#bioconductor"><i class="fa fa-check"></i><b>4.1</b> Bioconductor</a></li>
<li class="chapter" data-level="4.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#installing-biocmanager"><i class="fa fa-check"></i><b>4.2</b> Installing BiocManager</a></li>
<li class="chapter" data-level="4.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#the-ins-and-outs-of-package-installation"><i class="fa fa-check"></i><b>4.3</b> The ins and outs of package installation</a>
<ul>
<li class="chapter" data-level="4.3.1" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#updating-other-packages-when-downloading-a-package"><i class="fa fa-check"></i><b>4.3.1</b> Updating other packages when downloading a package</a></li>
<li class="chapter" data-level="4.3.2" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#packages-from-source"><i class="fa fa-check"></i><b>4.3.2</b> Packages “from source”</a></li>
<li class="chapter" data-level="4.3.3" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#more-on-angry-red-text"><i class="fa fa-check"></i><b>4.3.3</b> More on angry red text</a></li>
</ul></li>
<li class="chapter" data-level="4.4" data-path="installing-bioconductor.html"><a href="installing-bioconductor.html#actually-loading-a-package"><i class="fa fa-check"></i><b>4.4</b> Actually loading a package</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="basic-R.html"><a href="basic-R.html"><i class="fa fa-check"></i><b>5</b> A Brief introduction to R</a>
<ul>
<li class="chapter" data-level="5.1" data-path="basic-R.html"><a href="basic-R.html#vocabulary"><i class="fa fa-check"></i><b>5.1</b> Vocabulary</a></li>
<li class="chapter" data-level="5.2" data-path="basic-R.html"><a href="basic-R.html#r-functions"><i class="fa fa-check"></i><b>5.2</b> R functions</a></li>
<li class="chapter" data-level="5.3" data-path="basic-R.html"><a href="basic-R.html#interacting-with-r"><i class="fa fa-check"></i><b>5.3</b> Interacting with R</a></li>
<li class="chapter" data-level="5.4" data-path="basic-R.html"><a href="basic-R.html#variables-in-r"><i class="fa fa-check"></i><b>5.4</b> Variables in R</a>
<ul>
<li class="chapter" data-level="5.4.1" data-path="basic-R.html"><a href="basic-R.html#vectors"><i class="fa fa-check"></i><b>5.4.1</b> Vectors</a></li>
<li class="chapter" data-level="5.4.2" data-path="basic-R.html"><a href="basic-R.html#vector-indexing"><i class="fa fa-check"></i><b>5.4.2</b> Vector indexing</a></li>
<li class="chapter" data-level="5.4.3" data-path="basic-R.html"><a href="basic-R.html#character-vectors"><i class="fa fa-check"></i><b>5.4.3</b> Character vectors</a></li>
<li class="chapter" data-level="5.4.4" data-path="basic-R.html"><a href="basic-R.html#lists"><i class="fa fa-check"></i><b>5.4.4</b> Lists</a></li>
<li class="chapter" data-level="5.4.5" data-path="basic-R.html"><a href="basic-R.html#tables"><i class="fa fa-check"></i><b>5.4.5</b> Tables</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="basic-R.html"><a href="basic-R.html#arguments"><i class="fa fa-check"></i><b>5.5</b> Arguments</a></li>
<li class="chapter" data-level="5.6" data-path="basic-R.html"><a href="basic-R.html#help-files-with-help-and"><i class="fa fa-check"></i><b>5.6</b> Help files with <code>help()</code> and <code>?</code></a></li>
<li class="chapter" data-level="5.7" data-path="basic-R.html"><a href="basic-R.html#searching-for-functions-with-help.search-and-rsitesearch"><i class="fa fa-check"></i><b>5.7</b> Searching for functions with <code>help.search()</code> and <code>RSiteSearch()</code></a></li>
<li class="chapter" data-level="5.8" data-path="basic-R.html"><a href="basic-R.html#more-on-functions"><i class="fa fa-check"></i><b>5.8</b> More on functions</a>
<ul>
<li class="chapter" data-level="5.8.1" data-path="basic-R.html"><a href="basic-R.html#writing-your-own-functions"><i class="fa fa-check"></i><b>5.8.1</b> Writing your own functions</a></li>
</ul></li>
<li class="chapter" data-level="5.9" data-path="basic-R.html"><a href="basic-R.html#quiting-r"><i class="fa fa-check"></i><b>5.9</b> Quiting R</a></li>
<li class="chapter" data-level="5.10" data-path="basic-R.html"><a href="basic-R.html#links-and-further-reading"><i class="fa fa-check"></i><b>5.10</b> Links and Further Reading</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="vectors-in-R.html"><a href="vectors-in-R.html"><i class="fa fa-check"></i><b>6</b> A primer for working with vectors</a>
<ul>
<li class="chapter" data-level="6.1" data-path="vectors-in-R.html"><a href="vectors-in-R.html#preface-1"><i class="fa fa-check"></i><b>6.1</b> Preface</a></li>
<li class="chapter" data-level="6.2" data-path="vectors-in-R.html"><a href="vectors-in-R.html#vocab"><i class="fa fa-check"></i><b>6.2</b> Vocab</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="functions.html"><a href="functions.html"><i class="fa fa-check"></i><b>7</b> Functions</a>
<ul>
<li class="chapter" data-level="7.1" data-path="functions.html"><a href="functions.html#vectors-in-r"><i class="fa fa-check"></i><b>7.1</b> Vectors in R</a></li>
<li class="chapter" data-level="7.2" data-path="functions.html"><a href="functions.html#math-on-vectors"><i class="fa fa-check"></i><b>7.2</b> Math on vectors</a></li>
<li class="chapter" data-level="7.3" data-path="functions.html"><a href="functions.html#functions-on-vectors"><i class="fa fa-check"></i><b>7.3</b> Functions on vectors</a></li>
<li class="chapter" data-level="7.4" data-path="functions.html"><a href="functions.html#operations-with-two-vectors"><i class="fa fa-check"></i><b>7.4</b> Operations with two vectors</a></li>
<li class="chapter" data-level="7.5" data-path="functions.html"><a href="functions.html#subsetting-vectors"><i class="fa fa-check"></i><b>7.5</b> Subsetting vectors</a></li>
<li class="chapter" data-level="7.6" data-path="functions.html"><a href="functions.html#sequences-of-numbers"><i class="fa fa-check"></i><b>7.6</b> Sequences of numbers</a></li>
<li class="chapter" data-level="7.7" data-path="functions.html"><a href="functions.html#vectors-can-hold-numeric-or-character-data"><i class="fa fa-check"></i><b>7.7</b> Vectors can hold numeric or character data</a></li>
<li class="chapter" data-level="7.8" data-path="functions.html"><a href="functions.html#regular-expressions-can-modify-character-data"><i class="fa fa-check"></i><b>7.8</b> Regular expressions can modify character data</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="plotting-vectors.html"><a href="plotting-vectors.html"><i class="fa fa-check"></i><b>8</b> Plotting vectors in base R</a>
<ul>
<li class="chapter" data-level="8.1" data-path="plotting-vectors.html"><a href="plotting-vectors.html#preface-2"><i class="fa fa-check"></i><b>8.1</b> Preface</a></li>
<li class="chapter" data-level="8.2" data-path="plotting-vectors.html"><a href="plotting-vectors.html#plotting-numeric-data"><i class="fa fa-check"></i><b>8.2</b> Plotting numeric data</a></li>
<li class="chapter" data-level="8.3" data-path="plotting-vectors.html"><a href="plotting-vectors.html#other-plotting-packages"><i class="fa fa-check"></i><b>8.3</b> Other plotting packages</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="R-objects.html"><a href="R-objects.html"><i class="fa fa-check"></i><b>9</b> Intro to R objects</a>
<ul>
<li class="chapter" data-level="9.1" data-path="R-objects.html"><a href="R-objects.html#commands-used"><i class="fa fa-check"></i><b>9.1</b> Commands used</a></li>
<li class="chapter" data-level="9.2" data-path="R-objects.html"><a href="R-objects.html#r-objects"><i class="fa fa-check"></i><b>9.2</b> R Objects</a></li>
<li class="chapter" data-level="9.3" data-path="R-objects.html"><a href="R-objects.html#differences-between-objects"><i class="fa fa-check"></i><b>9.3</b> Differences between objects</a></li>
<li class="chapter" data-level="9.4" data-path="R-objects.html"><a href="R-objects.html#the-data"><i class="fa fa-check"></i><b>9.4</b> The Data</a></li>
<li class="chapter" data-level="9.5" data-path="R-objects.html"><a href="R-objects.html#the-assignment-operator---makes-object"><i class="fa fa-check"></i><b>9.5</b> The assignment operator “&lt;-” makes object</a>
<ul>
<li class="chapter" data-level="9.5.1" data-path="R-objects.html"><a href="R-objects.html#is"><i class="fa fa-check"></i><b>9.5.1</b> is()</a></li>
<li class="chapter" data-level="9.5.2" data-path="R-objects.html"><a href="R-objects.html#length"><i class="fa fa-check"></i><b>9.5.2</b> length()</a></li>
<li class="chapter" data-level="9.5.3" data-path="R-objects.html"><a href="R-objects.html#str"><i class="fa fa-check"></i><b>9.5.3</b> str()</a></li>
<li class="chapter" data-level="9.5.4" data-path="R-objects.html"><a href="R-objects.html#c"><i class="fa fa-check"></i><b>9.5.4</b> c()</a></li>
</ul></li>
<li class="chapter" data-level="9.6" data-path="R-objects.html"><a href="R-objects.html#debrief"><i class="fa fa-check"></i><b>9.6</b> Debrief</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html"><i class="fa fa-check"></i><b>10</b> Nucleic Acids</a>
<ul>
<li class="chapter" data-level="10.1" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#dna-and-rna"><i class="fa fa-check"></i><b>10.1</b> DNA and RNA</a></li>
<li class="chapter" data-level="10.2" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#dna-double-helix-structure"><i class="fa fa-check"></i><b>10.2</b> DNA Double-Helix Structure</a></li>
<li class="chapter" data-level="10.3" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#rna"><i class="fa fa-check"></i><b>10.3</b> RNA</a></li>
<li class="chapter" data-level="10.4" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#summary"><i class="fa fa-check"></i><b>10.4</b> Summary</a></li>
<li class="chapter" data-level="10.5" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#analysis-questions"><i class="fa fa-check"></i><b>10.5</b> Analysis Questions</a></li>
<li class="chapter" data-level="10.6" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#glossary"><i class="fa fa-check"></i><b>10.6</b> Glossary</a></li>
<li class="chapter" data-level="10.7" data-path="nucleic-acids-review.html"><a href="nucleic-acids-review.html#contributors-and-attributions"><i class="fa fa-check"></i><b>10.7</b> Contributors and Attributions</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="proteins-review.html"><a href="proteins-review.html"><i class="fa fa-check"></i><b>11</b> Proteins</a>
<ul>
<li class="chapter" data-level="11.0.1" data-path="proteins-review.html"><a href="proteins-review.html#skills-to-develop"><i class="fa fa-check"></i><b>11.0.1</b> Skills to develop:</a></li>
<li class="chapter" data-level="11.1" data-path="proteins-review.html"><a href="proteins-review.html#types-and-functions-of-proteins"><i class="fa fa-check"></i><b>11.1</b> Types and Functions of Proteins</a></li>
<li class="chapter" data-level="11.2" data-path="proteins-review.html"><a href="proteins-review.html#amino-acids"><i class="fa fa-check"></i><b>11.2</b> Amino Acids</a></li>
<li class="chapter" data-level="11.3" data-path="proteins-review.html"><a href="proteins-review.html#evolution-connection"><i class="fa fa-check"></i><b>11.3</b> Evolution Connection:</a></li>
<li class="chapter" data-level="11.4" data-path="proteins-review.html"><a href="proteins-review.html#protein-structure"><i class="fa fa-check"></i><b>11.4</b> Protein Structure</a>
<ul>
<li class="chapter" data-level="11.4.1" data-path="proteins-review.html"><a href="proteins-review.html#primary-structure"><i class="fa fa-check"></i><b>11.4.1</b> Primary Structure</a></li>
<li class="chapter" data-level="11.4.2" data-path="proteins-review.html"><a href="proteins-review.html#secondary-structure"><i class="fa fa-check"></i><b>11.4.2</b> Secondary Structure</a></li>
<li class="chapter" data-level="11.4.3" data-path="proteins-review.html"><a href="proteins-review.html#tertiary-structure"><i class="fa fa-check"></i><b>11.4.3</b> Tertiary Structure</a></li>
<li class="chapter" data-level="11.4.4" data-path="proteins-review.html"><a href="proteins-review.html#quaternary-structure"><i class="fa fa-check"></i><b>11.4.4</b> Quaternary Structure</a></li>
</ul></li>
<li class="chapter" data-level="11.5" data-path="proteins-review.html"><a href="proteins-review.html#denaturation-and-protein-folding"><i class="fa fa-check"></i><b>11.5</b> Denaturation and Protein Folding</a></li>
<li class="chapter" data-level="11.6" data-path="proteins-review.html"><a href="proteins-review.html#summary-1"><i class="fa fa-check"></i><b>11.6</b> Summary</a></li>
<li class="chapter" data-level="11.7" data-path="proteins-review.html"><a href="proteins-review.html#art-connections"><i class="fa fa-check"></i><b>11.7</b> Art Connections</a></li>
<li class="chapter" data-level="11.8" data-path="proteins-review.html"><a href="proteins-review.html#review-questions"><i class="fa fa-check"></i><b>11.8</b> Review Questions</a></li>
<li class="chapter" data-level="11.9" data-path="proteins-review.html"><a href="proteins-review.html#free-response"><i class="fa fa-check"></i><b>11.9</b> Free Response</a></li>
<li class="chapter" data-level="11.10" data-path="proteins-review.html"><a href="proteins-review.html#glossary-1"><i class="fa fa-check"></i><b>11.10</b> Glossary</a></li>
<li class="chapter" data-level="11.11" data-path="proteins-review.html"><a href="proteins-review.html#contributors-and-attributions-1"><i class="fa fa-check"></i><b>11.11</b> Contributors and Attributions</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="ncbi-the-national-center-for-biotechnology-information.html"><a href="ncbi-the-national-center-for-biotechnology-information.html"><i class="fa fa-check"></i><b>12</b> NCBI: The National Center for Biotechnology Information</a>
<ul>
<li class="chapter" data-level="12.1" data-path="ncbi-the-national-center-for-biotechnology-information.html"><a href="ncbi-the-national-center-for-biotechnology-information.html#key-concepts"><i class="fa fa-check"></i><b>12.1</b> Key concepts</a></li>
<li class="chapter" data-level="12.2" data-path="ncbi-the-national-center-for-biotechnology-information.html"><a href="ncbi-the-national-center-for-biotechnology-information.html#ncbi"><i class="fa fa-check"></i><b>12.2</b> NCBI</a></li>
<li class="chapter" data-level="12.3" data-path="ncbi-the-national-center-for-biotechnology-information.html"><a href="ncbi-the-national-center-for-biotechnology-information.html#genbank-sequence-database"><i class="fa fa-check"></i><b>12.3</b> GenBank sequence database</a></li>
<li class="chapter" data-level="12.4" data-path="ncbi-the-national-center-for-biotechnology-information.html"><a href="ncbi-the-national-center-for-biotechnology-information.html#pubmed-and-pubmed-central-article-database"><i class="fa fa-check"></i><b>12.4</b> PubMed and PubMed Central article database</a></li>
<li class="chapter" data-level="12.5" data-path="ncbi-the-national-center-for-biotechnology-information.html"><a href="ncbi-the-national-center-for-biotechnology-information.html#entrez"><i class="fa fa-check"></i><b>12.5</b> Entrez</a></li>
<li class="chapter" data-level="12.6" data-path="ncbi-the-national-center-for-biotechnology-information.html"><a href="ncbi-the-national-center-for-biotechnology-information.html#blast"><i class="fa fa-check"></i><b>12.6</b> BLAST</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html"><i class="fa fa-check"></i><b>13</b> Introduction to biological sequences databases</a>
<ul>
<li class="chapter" data-level="13.1" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#topics"><i class="fa fa-check"></i><b>13.1</b> Topics</a></li>
<li class="chapter" data-level="13.2" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#introduction"><i class="fa fa-check"></i><b>13.2</b> Introduction</a></li>
<li class="chapter" data-level="13.3" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#biological-sequence-databases"><i class="fa fa-check"></i><b>13.3</b> Biological sequence databases</a></li>
<li class="chapter" data-level="13.4" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#the-ncbi-sequence-database"><i class="fa fa-check"></i><b>13.4</b> The NCBI Sequence Database</a></li>
<li class="chapter" data-level="13.5" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#the-ncbi-sub-databases"><i class="fa fa-check"></i><b>13.5</b> The NCBI Sub-Databases</a></li>
<li class="chapter" data-level="13.6" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#ncbi-genbank-record-format"><i class="fa fa-check"></i><b>13.6</b> NCBI GenBank Record Format</a></li>
<li class="chapter" data-level="13.7" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#the-fasta-file-format"><i class="fa fa-check"></i><b>13.7</b> The FASTA file format</a></li>
<li class="chapter" data-level="13.8" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#refseq"><i class="fa fa-check"></i><b>13.8</b> RefSeq</a></li>
<li class="chapter" data-level="13.9" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#querying-the-ncbi-database"><i class="fa fa-check"></i><b>13.9</b> Querying the NCBI Database</a></li>
<li class="chapter" data-level="13.10" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#querying-the-ncbi-database-via-the-ncbi-website-for-reference"><i class="fa fa-check"></i><b>13.10</b> Querying the NCBI Database via the NCBI Website (for reference)</a></li>
<li class="chapter" data-level="13.11" data-path="introduction-to-biological-sequences-databases.html"><a href="introduction-to-biological-sequences-databases.html#example-finding-the-sequences-published-in-nature-460352-358-for-reference"><i class="fa fa-check"></i><b>13.11</b> Example: finding the sequences published in Nature 460:352-358 (for reference)</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="downloading-ncbi-sequence-data-by-hand.html"><a href="downloading-ncbi-sequence-data-by-hand.html"><i class="fa fa-check"></i><b>14</b> Downloading NCBI sequence data by hand</a>
<ul>
<li class="chapter" data-level="14.1" data-path="downloading-ncbi-sequence-data-by-hand.html"><a href="downloading-ncbi-sequence-data-by-hand.html#preface-3"><i class="fa fa-check"></i><b>14.1</b> Preface</a></li>
<li class="chapter" data-level="14.2" data-path="downloading-ncbi-sequence-data-by-hand.html"><a href="downloading-ncbi-sequence-data-by-hand.html#retrieving-genome-sequence-data-via-the-ncbi-website"><i class="fa fa-check"></i><b>14.2</b> Retrieving genome sequence data via the NCBI website</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html"><i class="fa fa-check"></i><b>15</b> Downloading sequences from UniProt by hand</a>
<ul>
<li class="chapter" data-level="15.1" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#vocab-1"><i class="fa fa-check"></i><b>15.1</b> Vocab</a></li>
<li class="chapter" data-level="15.2" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#downloading-protein-data-from-uniprot"><i class="fa fa-check"></i><b>15.2</b> Downloading Protein data from UniProt</a></li>
<li class="chapter" data-level="15.3" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#viewing-the-uniprot-webpage-for-a-protein-sequence"><i class="fa fa-check"></i><b>15.3</b> Viewing the UniProt webpage for a protein sequence</a>
<ul>
<li class="chapter" data-level="15.3.1" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#protein-function"><i class="fa fa-check"></i><b>15.3.1</b> Protein function</a></li>
<li class="chapter" data-level="15.3.2" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#protein-sequence-and-size"><i class="fa fa-check"></i><b>15.3.2</b> Protein sequence and size</a></li>
<li class="chapter" data-level="15.3.3" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#other-information"><i class="fa fa-check"></i><b>15.3.3</b> Other information</a></li>
</ul></li>
<li class="chapter" data-level="15.4" data-path="downloading-sequences-from-uniprot-by-hand.html"><a href="downloading-sequences-from-uniprot-by-hand.html#retrieving-a-uniprot-protein-sequence-via-the-uniprot-website"><i class="fa fa-check"></i><b>15.4</b> Retrieving a UniProt protein sequence via the UniProt website</a></li>
</ul></li>
<li class="chapter" data-level="16" data-path="introducingFASTA.html"><a href="introducingFASTA.html"><i class="fa fa-check"></i><b>16</b> Introducing FASTA Files</a>
<ul>
<li class="chapter" data-level="16.1" data-path="introducingFASTA.html"><a href="introducingFASTA.html#example-fasta-file"><i class="fa fa-check"></i><b>16.1</b> Example FASTA file</a></li>
<li class="chapter" data-level="16.2" data-path="introducingFASTA.html"><a href="introducingFASTA.html#multiple-sequences-in-a-single-fasta-file"><i class="fa fa-check"></i><b>16.2</b> Multiple sequences in a single FASTA file</a></li>
<li class="chapter" data-level="16.3" data-path="introducingFASTA.html"><a href="introducingFASTA.html#multiple-sequence-alignments-can-be-stored-in-fasta-format"><i class="fa fa-check"></i><b>16.3</b> Multiple sequence alignments can be stored in FASTA format</a></li>
<li class="chapter" data-level="16.4" data-path="introducingFASTA.html"><a href="introducingFASTA.html#fastq-format"><i class="fa fa-check"></i><b>16.4</b> FASTQ Format</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="a-complete-bioinformatics-workflow-in-r.html"><a href="a-complete-bioinformatics-workflow-in-r.html"><i class="fa fa-check"></i><b>17</b> A complete bioinformatics workflow in R</a></li>
<li class="chapter" data-level="18" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html"><i class="fa fa-check"></i><b>18</b> “Worked example: Building a phylogeny in R”</a>
<ul>
<li class="chapter" data-level="18.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#introduction-1"><i class="fa fa-check"></i><b>18.1</b> Introduction</a>
<ul>
<li class="chapter" data-level="18.1.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#vocab-2"><i class="fa fa-check"></i><b>18.1.1</b> Vocab</a></li>
</ul></li>
<li class="chapter" data-level="18.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#software-preliminaires"><i class="fa fa-check"></i><b>18.2</b> Software Preliminaires</a>
<ul>
<li class="chapter" data-level="18.2.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#r-functions-1"><i class="fa fa-check"></i><b>18.2.1</b> R functions</a></li>
<li class="chapter" data-level="18.2.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#download-necessary-packages"><i class="fa fa-check"></i><b>18.2.2</b> Download necessary packages</a></li>
<li class="chapter" data-level="18.2.3" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#load-packages-into-memory"><i class="fa fa-check"></i><b>18.2.3</b> Load packages into memory</a></li>
</ul></li>
<li class="chapter" data-level="18.3" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#downloading-macro-molecular-sequences"><i class="fa fa-check"></i><b>18.3</b> Downloading macro-molecular sequences</a></li>
<li class="chapter" data-level="18.4" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#prepping-macromolecular-sequences"><i class="fa fa-check"></i><b>18.4</b> Prepping macromolecular sequences</a></li>
<li class="chapter" data-level="18.5" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#aligning-sequences"><i class="fa fa-check"></i><b>18.5</b> Aligning sequences</a></li>
<li class="chapter" data-level="18.6" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#the-shroom-family-of-genes"><i class="fa fa-check"></i><b>18.6</b> The shroom family of genes</a></li>
<li class="chapter" data-level="18.7" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#downloading-multiple-sequences"><i class="fa fa-check"></i><b>18.7</b> Downloading multiple sequences</a></li>
<li class="chapter" data-level="18.8" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#multiple-sequence-alignment"><i class="fa fa-check"></i><b>18.8</b> Multiple sequence alignment</a>
<ul>
<li class="chapter" data-level="18.8.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#building-an-multiple-sequence-alignment-msa"><i class="fa fa-check"></i><b>18.8.1</b> Building an Multiple Sequence Alignment (MSA)</a></li>
<li class="chapter" data-level="18.8.2" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#viewing-an-msa"><i class="fa fa-check"></i><b>18.8.2</b> Viewing an MSA</a></li>
</ul></li>
<li class="chapter" data-level="18.9" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#a-subset-of-sequences"><i class="fa fa-check"></i><b>18.9</b> A subset of sequences</a></li>
<li class="chapter" data-level="18.10" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#genetic-distances-of-sequence-in-subset"><i class="fa fa-check"></i><b>18.10</b> Genetic distances of sequence in subset</a></li>
<li class="chapter" data-level="18.11" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#phylognetic-trees-of-subset-sequences-finally"><i class="fa fa-check"></i><b>18.11</b> Phylognetic trees of subset sequences (finally!)</a>
<ul>
<li class="chapter" data-level="18.11.1" data-path="worked-example-building-a-phylogeny-in-r.html"><a href="worked-example-building-a-phylogeny-in-r.html#plotting-phylogenetic-trees"><i class="fa fa-check"></i><b>18.11.1</b> Plotting phylogenetic trees</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="19" data-path="simple-for-loop-example.html"><a href="simple-for-loop-example.html"><i class="fa fa-check"></i><b>19</b> Simple for() loop example</a>
<ul>
<li class="chapter" data-level="19.1" data-path="simple-for-loop-example.html"><a href="simple-for-loop-example.html#key-functions-terms"><i class="fa fa-check"></i><b>19.1</b> Key functions / terms</a></li>
</ul></li>
<li class="chapter" data-level="20" data-path="phylogenetic-tree-example-using-neighbor-joining.html"><a href="phylogenetic-tree-example-using-neighbor-joining.html"><i class="fa fa-check"></i><b>20</b> Phylogenetic tree example using neighbor joining</a>
<ul>
<li class="chapter" data-level="20.1" data-path="phylogenetic-tree-example-using-neighbor-joining.html"><a href="phylogenetic-tree-example-using-neighbor-joining.html#key-vocab-concepts-not-exhaustive"><i class="fa fa-check"></i><b>20.1</b> Key vocab / concepts (not exhaustive)</a></li>
</ul></li>
<li class="chapter" data-level="21" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html"><i class="fa fa-check"></i><b>21</b> Downloading DNA sequences as FASTA files in R</a>
<ul>
<li class="chapter" data-level="21.0.1" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#functions-1"><i class="fa fa-check"></i><b>21.0.1</b> Functions</a></li>
<li class="chapter" data-level="21.0.2" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#softwarewebsites"><i class="fa fa-check"></i><b>21.0.2</b> Software/websites</a></li>
<li class="chapter" data-level="21.0.3" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#r-vocabulary"><i class="fa fa-check"></i><b>21.0.3</b> R vocabulary</a></li>
<li class="chapter" data-level="21.0.4" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#file-types"><i class="fa fa-check"></i><b>21.0.4</b> File types</a></li>
<li class="chapter" data-level="21.0.5" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#bioinformatics-vocabulary"><i class="fa fa-check"></i><b>21.0.5</b> Bioinformatics vocabulary</a></li>
<li class="chapter" data-level="21.1" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#learning-objectives"><i class="fa fa-check"></i><b>21.1</b> Learning objectives</a>
<ul>
<li class="chapter" data-level="21.1.1" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#organisms-and-sequence-accessions"><i class="fa fa-check"></i><b>21.1.1</b> Organisms and Sequence accessions</a></li>
<li class="chapter" data-level="21.1.2" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#preliminaries"><i class="fa fa-check"></i><b>21.1.2</b> Preliminaries</a></li>
</ul></li>
<li class="chapter" data-level="21.2" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#dna-sequence-statistics-part-1"><i class="fa fa-check"></i><b>21.2</b> DNA Sequence Statistics: Part 1</a>
<ul>
<li class="chapter" data-level="21.2.1" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#using-r-for-bioinformatics"><i class="fa fa-check"></i><b>21.2.1</b> Using R for Bioinformatics</a></li>
<li class="chapter" data-level="21.2.2" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#r-packages-for-bioinformatics-bioconductor-and-seqinr"><i class="fa fa-check"></i><b>21.2.2</b> R packages for bioinformatics: Bioconductor and SeqinR</a></li>
<li class="chapter" data-level="21.2.3" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#fasta-file-format"><i class="fa fa-check"></i><b>21.2.3</b> FASTA file format</a></li>
<li class="chapter" data-level="21.2.4" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#the-ncbi-sequence-database-1"><i class="fa fa-check"></i><b>21.2.4</b> The NCBI sequence database</a></li>
<li class="chapter" data-level="21.2.5" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#retrieving-genome-sequence-data-using-rentrez"><i class="fa fa-check"></i><b>21.2.5</b> Retrieving genome sequence data using rentrez</a></li>
</ul></li>
<li class="chapter" data-level="21.3" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#saving-fasta-files"><i class="fa fa-check"></i><b>21.3</b> Saving FASTA files</a></li>
<li class="chapter" data-level="21.4" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#next-steps"><i class="fa fa-check"></i><b>21.4</b> Next steps</a></li>
<li class="chapter" data-level="21.5" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#exercises"><i class="fa fa-check"></i><b>21.5</b> Exercises</a></li>
<li class="chapter" data-level="21.6" data-path="download-FASTA-inR.html"><a href="download-FASTA-inR.html#review-questions-1"><i class="fa fa-check"></i><b>21.6</b> Review questions</a></li>
</ul></li>
<li class="chapter" data-level="22" data-path="cleaning-and-preparing-fasta-files-for-analysis-in-r.html"><a href="cleaning-and-preparing-fasta-files-for-analysis-in-r.html"><i class="fa fa-check"></i><b>22</b> Cleaning and preparing FASTA files for analysis in R</a>
<ul>
<li class="chapter" data-level="22.1" data-path="cleaning-and-preparing-fasta-files-for-analysis-in-r.html"><a href="cleaning-and-preparing-fasta-files-for-analysis-in-r.html#preliminaries-1"><i class="fa fa-check"></i><b>22.1</b> Preliminaries</a></li>
<li class="chapter" data-level="22.2" data-path="cleaning-and-preparing-fasta-files-for-analysis-in-r.html"><a href="cleaning-and-preparing-fasta-files-for-analysis-in-r.html#convert-fasta-sequence-to-an-r-variable"><i class="fa fa-check"></i><b>22.2</b> Convert FASTA sequence to an R variable</a>
<ul>
<li class="chapter" data-level="22.2.1" data-path="cleaning-and-preparing-fasta-files-for-analysis-in-r.html"><a href="cleaning-and-preparing-fasta-files-for-analysis-in-r.html#removing-unwanted-characters"><i class="fa fa-check"></i><b>22.2.1</b> Removing unwanted characters</a></li>
<li class="chapter" data-level="22.2.2" data-path="cleaning-and-preparing-fasta-files-for-analysis-in-r.html"><a href="cleaning-and-preparing-fasta-files-for-analysis-in-r.html#spliting-unbroken-strings-in-character-vectors"><i class="fa fa-check"></i><b>22.2.2</b> Spliting unbroken strings in character vectors</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="23" data-path="dna-descriptive-statics---part-1.html"><a href="dna-descriptive-statics---part-1.html"><i class="fa fa-check"></i><b>23</b> DNA descriptive statics - Part 1</a>
<ul>
<li class="chapter" data-level="23.1" data-path="dna-descriptive-statics---part-1.html"><a href="dna-descriptive-statics---part-1.html#preface-4"><i class="fa fa-check"></i><b>23.1</b> Preface</a></li>
<li class="chapter" data-level="23.2" data-path="dna-descriptive-statics---part-1.html"><a href="dna-descriptive-statics---part-1.html#introduction-2"><i class="fa fa-check"></i><b>23.2</b> Introduction</a></li>
<li class="chapter" data-level="23.3" data-path="dna-descriptive-statics---part-1.html"><a href="dna-descriptive-statics---part-1.html#vocabulary-1"><i class="fa fa-check"></i><b>23.3</b> Vocabulary</a></li>
<li class="chapter" data-level="23.4" data-path="dna-descriptive-statics---part-1.html"><a href="dna-descriptive-statics---part-1.html#functions-2"><i class="fa fa-check"></i><b>23.4</b> Functions</a></li>
<li class="chapter" data-level="23.5" data-path="dna-descriptive-statics---part-1.html"><a href="dna-descriptive-statics---part-1.html#learning-objectives-1"><i class="fa fa-check"></i><b>23.5</b> Learning objectives</a></li>
<li class="chapter" data-level="23.6" data-path="dna-descriptive-statics---part-1.html"><a href="dna-descriptive-statics---part-1.html#preliminaries-2"><i class="fa fa-check"></i><b>23.6</b> Preliminaries</a></li>
<li class="chapter" data-level="23.7" data-path="dna-descriptive-statics---part-1.html"><a href="dna-descriptive-statics---part-1.html#converting-dna-from-fasta-format"><i class="fa fa-check"></i><b>23.7</b> Converting DNA from FASTA format</a></li>
<li class="chapter" data-level="23.8" data-path="dna-descriptive-statics---part-1.html"><a href="dna-descriptive-statics---part-1.html#length-of-a-dna-sequence"><i class="fa fa-check"></i><b>23.8</b> Length of a DNA sequence</a>
<ul>
<li class="chapter" data-level="23.8.1" data-path="dna-descriptive-statics---part-1.html"><a href="dna-descriptive-statics---part-1.html#base-composition-of-a-dna-sequence"><i class="fa fa-check"></i><b>23.8.1</b> Base composition of a DNA sequence</a></li>
<li class="chapter" data-level="23.8.2" data-path="dna-descriptive-statics---part-1.html"><a href="dna-descriptive-statics---part-1.html#gc-content-of-dna"><i class="fa fa-check"></i><b>23.8.2</b> GC Content of DNA</a></li>
<li class="chapter" data-level="23.8.3" data-path="dna-descriptive-statics---part-1.html"><a href="dna-descriptive-statics---part-1.html#dna-words"><i class="fa fa-check"></i><b>23.8.3</b> DNA words</a></li>
<li class="chapter" data-level="23.8.4" data-path="dna-descriptive-statics---part-1.html"><a href="dna-descriptive-statics---part-1.html#summary-2"><i class="fa fa-check"></i><b>23.8.4</b> Summary</a></li>
</ul></li>
<li class="chapter" data-level="23.9" data-path="dna-descriptive-statics---part-1.html"><a href="dna-descriptive-statics---part-1.html#acknowledgements"><i class="fa fa-check"></i><b>23.9</b> Acknowledgements</a>
<ul>
<li class="chapter" data-level="23.9.1" data-path="dna-descriptive-statics---part-1.html"><a href="dna-descriptive-statics---part-1.html#license"><i class="fa fa-check"></i><b>23.9.1</b> License</a></li>
<li class="chapter" data-level="23.9.2" data-path="dna-descriptive-statics---part-1.html"><a href="dna-descriptive-statics---part-1.html#exercises-1"><i class="fa fa-check"></i><b>23.9.2</b> Exercises</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path=""><a href="#part-appendices"><i class="fa fa-check"></i>(PART) Appendices</a>
<ul>
<li><a href="#section"></a></li>
</ul></li>
<li class="chapter" data-level="" data-path="appendix-02-getting-access-to-r.html"><a href="appendix-02-getting-access-to-r.html"><i class="fa fa-check"></i>Appendix 02: Getting access to R</a>
<ul>
<li class="chapter" data-level="23.10" data-path="appendix-02-getting-access-to-r.html"><a href="appendix-02-getting-access-to-r.html#getting-started-with-r-and-rstudio"><i class="fa fa-check"></i><b>23.10</b> Getting Started With R and RStudio</a>
<ul>
<li class="chapter" data-level="23.10.1" data-path="appendix-02-getting-access-to-r.html"><a href="appendix-02-getting-access-to-r.html#rstudio-cloud"><i class="fa fa-check"></i><b>23.10.1</b> RStudio Cloud</a></li>
<li class="chapter" data-level="23.10.2" data-path="appendix-02-getting-access-to-r.html"><a href="appendix-02-getting-access-to-r.html#getting-r-onto-your-own-computer"><i class="fa fa-check"></i><b>23.10.2</b> Getting R onto your own computer</a></li>
<li class="chapter" data-level="23.10.3" data-path="appendix-02-getting-access-to-r.html"><a href="appendix-02-getting-access-to-r.html#getting-rstudio-onto-your-computer"><i class="fa fa-check"></i><b>23.10.3</b> Getting RStudio onto your computer</a></li>
<li class="chapter" data-level="23.10.4" data-path="appendix-02-getting-access-to-r.html"><a href="appendix-02-getting-access-to-r.html#keep-r-and-rstudio-current"><i class="fa fa-check"></i><b>23.10.4</b> Keep R and RStudio current</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html"><i class="fa fa-check"></i>Getting started with R itself (or not)</a>
<ul>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#vocabulary-2"><i class="fa fa-check"></i>Vocabulary</a></li>
<li class="chapter" data-level="" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#r-commands"><i class="fa fa-check"></i>R commands</a></li>
<li class="chapter" data-level="23.11" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#help"><i class="fa fa-check"></i><b>23.11</b> Help!</a>
<ul>
<li class="chapter" data-level="23.11.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-r"><i class="fa fa-check"></i><b>23.11.1</b> Getting “help” from R</a></li>
<li class="chapter" data-level="23.11.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-the-internet"><i class="fa fa-check"></i><b>23.11.2</b> Getting help from the internet</a></li>
<li class="chapter" data-level="23.11.3" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-online-forums"><i class="fa fa-check"></i><b>23.11.3</b> Getting help from online forums</a></li>
<li class="chapter" data-level="23.11.4" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#getting-help-from-twitter"><i class="fa fa-check"></i><b>23.11.4</b> Getting help from twitter</a></li>
</ul></li>
<li class="chapter" data-level="23.12" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#other-features-of-rstudio"><i class="fa fa-check"></i><b>23.12</b> Other features of RStudio</a>
<ul>
<li class="chapter" data-level="23.12.1" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#ajusting-pane-the-layout"><i class="fa fa-check"></i><b>23.12.1</b> Ajusting pane the layout</a></li>
<li class="chapter" data-level="23.12.2" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#adjusting-size-of-windows"><i class="fa fa-check"></i><b>23.12.2</b> Adjusting size of windows</a></li>
</ul></li>
<li class="chapter" data-level="23.13" data-path="getting-started-with-r-itself-or-not.html"><a href="getting-started-with-r-itself-or-not.html#practice-optional"><i class="fa fa-check"></i><b>23.13</b> Practice (OPTIONAL)</a></li>
</ul></li>
<li class="chapter" data-level="24" data-path="logarithms-in-r.html"><a href="logarithms-in-r.html"><i class="fa fa-check"></i><b>24</b> Logarithms in R</a></li>
<li class="divider"></li>
<li><a href="https://brouwern.github.io/lbrb/" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">A Little Book of R for Bioinformatics 2.0</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="worked-example-building-a-phylogeny-in-r" class="section level1" number="18">
<h1><span class="header-section-number">Chapter 18</span> “Worked example: Building a phylogeny in R”</h1>
<div id="introduction-1" class="section level2" number="18.1">
<h2><span class="header-section-number">18.1</span> Introduction</h2>
<p>Phylogenies play an important role in computational biology and bioinformatics. Phylogenetics itself is an obligatly computational field that only began rapid growth when computational power allowed the many algorithms it relies on to be done rapidly. Phylogeneies of species, genes and proteins are used to address many biological issues, including</p>
<ul>
<li>Patterns of protein evolution</li>
<li>Origin and evolution of phenotypic traits</li>
<li>Origin and progression of epidemics</li>
<li>Origin of evolution of diseases (e.g., zooenoses)</li>
<li>Prediction of protein function from its sequence</li>
<li>… and many more</li>
</ul>
<p>The actual building of a phylogeny is a computationally intensive task; moreover, there are many bioinformatics and computational tasks the precede the construction of a phylogeny:</p>
<ul>
<li>genome sequencing and assembly</li>
<li>computational gene prediction and annotation</li>
<li>database searching and results screening</li>
<li>pairwise sequence alignment</li>
<li>data organization and cleaning</li>
<li>multiple sequence alignment</li>
<li>evaluation and validation of alignment accuracy</li>
</ul>
<p>Once all of these steps have been carried out, the building of a phylogeny involves</p>
<ul>
<li>picking a model of sequence evolution or other description of evolution</li>
<li>picking a statistical approach to tree construction</li>
<li>evaluating uncertainty in the final tree</li>
</ul>
<p>In this chapter we will work through many of these steps. In most cases we will pick the easiest or fastest option; in later chapters we will unpack the various options. This chapter is written as an interactive R session. You can follow along by opening the .Rmd file of the chapter or typing the appropriate commands into your own script. I assume that all the necessary packages have been installed and they only need to be loaded into R using the <em>library()</em> command.</p>
<p>This lesson walks you through and entire workflow for a bioinformatics project, including</p>
<!-- check all of this -->
<ol style="list-style-type: decimal">
<li>obtaining FASTA sequences</li>
<li>cleaning sequences</li>
<li>creating alignments</li>
<li>creating a distance matrix</li>
<li>building a phylogenetic tree</li>
</ol>
<p>We’ll examine the Shroom family of genes, which produces Shroom proteins essential for tissue formation in many multicellular eukaryotes, including neural tube formation in vertebrates. We’ll examine shroom in several very different organism, including humans, mice and sea urchins. There is more than one type of shroom in vertebrates, and we’ll also look at two different Shroom genes: shroom 1 and shroom 2.</p>
<p>This lesson draws on skills from previous sections of the book, but is written to act as a independent summary of these activities. There is therefore review of key aspects of R and bioinformatics throughout it.</p>
<div id="vocab-2" class="section level3" number="18.1.1">
<h3><span class="header-section-number">18.1.1</span> Vocab</h3>
</div>
</div>
<div id="software-preliminaires" class="section level2" number="18.2">
<h2><span class="header-section-number">18.2</span> Software Preliminaires</h2>
<ul>
<li>argument</li>
<li>function</li>
<li>list</li>
<li>named list</li>
<li>vector</li>
<li>named vector</li>
<li>for() loop</li>
<li>R console</li>
</ul>
<div id="r-functions-1" class="section level3" number="18.2.1">
<h3><span class="header-section-number">18.2.1</span> R functions</h3>
<ul>
<li>library()</li>
<li>round()</li>
<li>plot()</li>
<li>mtext()</li>
<li>nchar()</li>
<li>rentrez::entrez_fetch()</li>
<li>compbio4all::entrez_fetch_list()</li>
<li>compbio4all::print_msa() (Coghlan 2011)</li>
<li>Biostrings::AAStringSet()</li>
<li>msa::msa()</li>
<li>msa::msaConvert()</li>
<li>msa::msaPrettyPrint()</li>
<li>seqinr::dist.alignment()</li>
<li>ape::nj()</li>
</ul>
<p>A few things need to be done to get started with our R session.</p>
</div>
<div id="download-necessary-packages" class="section level3" number="18.2.2">
<h3><span class="header-section-number">18.2.2</span> Download necessary packages</h3>
<p>Many <em>R</em> sessions begin by downloading necessary software packages to augment <em>R’s</em> functionality.</p>
<p>If you don’t have them already, you’ll need the following packages from CRAN:</p>
<ol style="list-style-type: decimal">
<li><code>ape</code></li>
<li><code>seqinr</code></li>
<li><code>rentrez</code></li>
<li><code>devtools</code></li>
</ol>
<p>The CRAN packages can be loaded with <code>install.packages()</code>.</p>
<p>You’ll also need these packages from Bioconductor:</p>
<ol style="list-style-type: decimal">
<li><code>msa</code></li>
<li><code>Biostrings</code></li>
</ol>
<p>For installing packages from Bioconductor, see the chapter at the beginning of this book on this process.</p>
<p>Finally, you’ll need these package from GitHub</p>
<ol style="list-style-type: decimal">
<li><code>compbio4all</code></li>
<li><code>ggmsa</code></li>
</ol>
<p>To install packages from GitHub you can use the code <code>devtools::install_github("brouwern/compbio4all")</code> and <code>devtools::install_github("YuLab-SMU/ggmsa")</code></p>
<p>This code chunk downloads most of these packages. The code has been commented out because it only needs to be run once.</p>
<div class="sourceCode" id="cb170"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb170-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Install needed packages</span></span>
<span id="cb170-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-3" aria-hidden="true" tabindex="-1"></a><span class="do">## CRAN PACKAGES</span></span>
<span id="cb170-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-4" aria-hidden="true" tabindex="-1"></a><span class="co"># downloaded using install.packages()</span></span>
<span id="cb170-5"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-5" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages(&quot;rentrez&quot;,dependencies = TRUE)</span></span>
<span id="cb170-6"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-6" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages(&quot;devtools&quot;)</span></span>
<span id="cb170-7"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-7" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages(&quot;ape&quot;)</span></span>
<span id="cb170-8"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-8" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages(&quot;seqinr&quot;)</span></span>
<span id="cb170-9"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-10"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-10" aria-hidden="true" tabindex="-1"></a><span class="do">### BiocManager - CRAN package to download </span></span>
<span id="cb170-11"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-11" aria-hidden="true" tabindex="-1"></a><span class="do">##### Bioconductor packages</span></span>
<span id="cb170-12"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-12" aria-hidden="true" tabindex="-1"></a><span class="co"># requires BiocManager</span></span>
<span id="cb170-13"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-13" aria-hidden="true" tabindex="-1"></a><span class="co"># install.packages(&quot;BiocManager&quot;)</span></span>
<span id="cb170-14"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-15"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-15" aria-hidden="true" tabindex="-1"></a><span class="do">## BioConductor packages</span></span>
<span id="cb170-16"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-16" aria-hidden="true" tabindex="-1"></a><span class="do">### downloaded with BiocManager::install(), NOT install.packages()</span></span>
<span id="cb170-17"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-17" aria-hidden="true" tabindex="-1"></a><span class="co"># BiocManager::install(&quot;msa&quot;)</span></span>
<span id="cb170-18"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-18" aria-hidden="true" tabindex="-1"></a><span class="co"># BiocManager::install(&quot;Biostrings&quot;)</span></span>
<span id="cb170-19"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-20"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb170-21"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-21" aria-hidden="true" tabindex="-1"></a><span class="do">## GitHub packages</span></span>
<span id="cb170-22"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-22" aria-hidden="true" tabindex="-1"></a><span class="do">### requires devtools package and its function install_github()</span></span>
<span id="cb170-23"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-23" aria-hidden="true" tabindex="-1"></a><span class="co"># library(devtools)</span></span>
<span id="cb170-24"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-24" aria-hidden="true" tabindex="-1"></a><span class="co"># devtools::install_github(&quot;brouwern/combio4all&quot;)</span></span>
<span id="cb170-25"><a href="worked-example-building-a-phylogeny-in-r.html#cb170-25" aria-hidden="true" tabindex="-1"></a><span class="co"># devtools::install_github(&quot;YuLab-SMU/ggmsa&quot;)</span></span></code></pre></div>
</div>
<div id="load-packages-into-memory" class="section level3" number="18.2.3">
<h3><span class="header-section-number">18.2.3</span> Load packages into memory</h3>
<p>We now need to load up all our bioinformatics and phylogenetics software into R. This is done with the <code>library()</code> command.</p>
<p>To run this code just click on the sideways green triangle all the way to the right of the code.</p>
<p>NOTE: You’ll likely see some red code appear on your screen. No worries, totally normal!</p>
<div class="sourceCode" id="cb171"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb171-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb171-1" aria-hidden="true" tabindex="-1"></a><span class="co"># github packages</span></span>
<span id="cb171-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb171-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(compbio4all)</span>
<span id="cb171-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb171-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggmsa)</span>
<span id="cb171-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb171-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb171-5"><a href="worked-example-building-a-phylogeny-in-r.html#cb171-5" aria-hidden="true" tabindex="-1"></a><span class="co"># CRAN packages</span></span>
<span id="cb171-6"><a href="worked-example-building-a-phylogeny-in-r.html#cb171-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rentrez)</span>
<span id="cb171-7"><a href="worked-example-building-a-phylogeny-in-r.html#cb171-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(seqinr)</span>
<span id="cb171-8"><a href="worked-example-building-a-phylogeny-in-r.html#cb171-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb171-9"><a href="worked-example-building-a-phylogeny-in-r.html#cb171-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb171-10"><a href="worked-example-building-a-phylogeny-in-r.html#cb171-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Bioconductor packages</span></span>
<span id="cb171-11"><a href="worked-example-building-a-phylogeny-in-r.html#cb171-11" aria-hidden="true" tabindex="-1"></a><span class="do">## msa</span></span>
<span id="cb171-12"><a href="worked-example-building-a-phylogeny-in-r.html#cb171-12" aria-hidden="true" tabindex="-1"></a><span class="do">### The msa package is having problems on some platforms</span></span>
<span id="cb171-13"><a href="worked-example-building-a-phylogeny-in-r.html#cb171-13" aria-hidden="true" tabindex="-1"></a><span class="do">### You can skip the msa steps if necessary.  The msa output</span></span>
<span id="cb171-14"><a href="worked-example-building-a-phylogeny-in-r.html#cb171-14" aria-hidden="true" tabindex="-1"></a><span class="do">### is used to make a distance matrix and then phylogenetics trees,</span></span>
<span id="cb171-15"><a href="worked-example-building-a-phylogeny-in-r.html#cb171-15" aria-hidden="true" tabindex="-1"></a><span class="do">### but I provide code to build the matrix by hand so</span></span>
<span id="cb171-16"><a href="worked-example-building-a-phylogeny-in-r.html#cb171-16" aria-hidden="true" tabindex="-1"></a><span class="do">### you can proceed even if msa doesn&#39;t work for you.</span></span>
<span id="cb171-17"><a href="worked-example-building-a-phylogeny-in-r.html#cb171-17" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(msa)</span>
<span id="cb171-18"><a href="worked-example-building-a-phylogeny-in-r.html#cb171-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb171-19"><a href="worked-example-building-a-phylogeny-in-r.html#cb171-19" aria-hidden="true" tabindex="-1"></a><span class="do">## Biostrings</span></span>
<span id="cb171-20"><a href="worked-example-building-a-phylogeny-in-r.html#cb171-20" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Biostrings)</span></code></pre></div>
</div>
</div>
<div id="downloading-macro-molecular-sequences" class="section level2" number="18.3">
<h2><span class="header-section-number">18.3</span> Downloading macro-molecular sequences</h2>
<p>We’re going to explore some sequences. First we need to download them. To do this we’ll use a function, <code>entrez_fretch()</code>, which accesses the <a href="ncbi.nlm.nih.gov/search/"><strong>Entrez</strong></a> system of database (ncbi.nlm.nih.gov/search/). This function is from the <code>rentrez</code> package, which stands for “R-Entrez.”</p>
<p>We need to tell <code>entrez_fetch()</code> several things</p>
<ol style="list-style-type: decimal">
<li><code>db = ...</code> the type of entrez database.<br />
</li>
<li><code>id = ...</code> the <strong>accession</strong> (ID) number of the sequence</li>
<li><code>rettype = ...</code> file type what we want the function to return.</li>
</ol>
<p>Formally, these things are called <strong>arguments</strong> by <em>R</em>.</p>
<p>We’ll use these settings:</p>
<ol style="list-style-type: decimal">
<li><code>db = "protein"</code> to access the Entrez database of protein sequences<br />
</li>
<li><code>rettype = "fasta"</code>, which is a standard file format for nucleic acid and protein sequences</li>
</ol>
<p>We’ll set <code>id = ...</code> to sequences whose <strong>accession numbers</strong> are:</p>
<ol style="list-style-type: decimal">
<li>NP_065910: Human shroom 3</li>
<li>AAF13269: Mouse shroom 3a</li>
<li>CAA58534: Human shroom 2</li>
<li>XP_783573: Sea urchin shroom</li>
</ol>
<p>There are two highly conserved regions of shroom 3
1. ASD 1: aa 884 to aa 1062 in hShroom3
1. ASD 2: aa 1671 to aa 1955 in hShroom3</p>
<p>Normally we’d have to download these sequences by hand through pointing and clicking on GeneBank records on the NCBI website. In <em>R</em> we can do it automatically; this might take a second.</p>
<p>All the code needed is this:</p>
<div class="sourceCode" id="cb172"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb172-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb172-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Human shroom 3 (H. sapiens)</span></span>
<span id="cb172-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb172-2" aria-hidden="true" tabindex="-1"></a>hShroom3 <span class="ot">&lt;-</span> rentrez<span class="sc">::</span><span class="fu">entrez_fetch</span>(<span class="at">db =</span> <span class="st">&quot;protein&quot;</span>, </span>
<span id="cb172-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb172-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">id =</span> <span class="st">&quot;NP_065910&quot;</span>, </span>
<span id="cb172-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb172-4" aria-hidden="true" tabindex="-1"></a>                          <span class="at">rettype =</span> <span class="st">&quot;fasta&quot;</span>)</span></code></pre></div>
<p>Note: if you aren’t connected to wifi or the server itself is having problem, then when you run this code you may get an error like this:
“Quitting from lines 244-259 (23-MSA-walkthrough-shroom.Rmd)
Error: HTTP failure: 500
{”error”:“error forwarding request”,“api-key”:“71.182.228.80”,“type”:“ip”,
“status”:“ok”}
Execution halted”</p>
<p>You may have to try again later to knit the code.</p>
<p>The output is in FASTA format. We can look at the raw output by calling up the object we created. This will run as single continuous string of characters without line breaks</p>
<div class="sourceCode" id="cb173"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb173-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb173-1" aria-hidden="true" tabindex="-1"></a>hShroom3</span></code></pre></div>
<pre><code>## [1] &quot;&gt;NP_065910.3 protein Shroom3 [Homo sapiens]\nMMRTTEDFHKPSATLNSNTATKGRYIYLEAFLEGGAPWGFTLKGGLEHGEPLIISKVEEGGKADTLSSKL\nQAGDEVVHINEVTLSSSRKEAVSLVKGSYKTLRLVVRRDVCTDPGHADTGASNFVSPEHLTSGPQHRKAA\nWSGGVKLRLKHRRSEPAGRPHSWHTTKSGEKQPDASMMQISQGMIGPPWHQSYHSSSSTSDLSNYDHAYL\nRRSPDQCSSQGSMESLEPSGAYPPCHLSPAKSTGSIDQLSHFHNKRDSAYSSFSTSSSILEYPHPGISGR\nERSGSMDNTSARGGLLEGMRQADIRYVKTVYDTRRGVSAEYEVNSSALLLQGREARASANGQGYDKWSNI\nPRGKGVPPPSWSQQCPSSLETATDNLPPKVGAPLPPARSDSYAAFRHRERPSSWSSLDQKRLCRPQANSL\nGSLKSPFIEEQLHTVLEKSPENSPPVKPKHNYTQKAQPGQPLLPTSIYPVPSLEPHFAQVPQPSVSSNGM\nLYPALAKESGYIAPQGACNKMATIDENGNQNGSGRPGFAFCQPLEHDLLSPVEKKPEATAKYVPSKVHFC\nSVPENEEDASLKRHLTPPQGNSPHSNERKSTHSNKPSSHPHSLKCPQAQAWQAGEDKRSSRLSEPWEGDF\nQEDHNANLWRRLEREGLGQSLSGNFGKTKSAFSSLQNIPESLRRHSSLELGRGTQEGYPGGRPTCAVNTK\nAEDPGRKAAPDLGSHLDRQVSYPRPEGRTGASASFNSTDPSPEEPPAPSHPHTSSLGRRGPGPGSASALQ\nGFQYGKPHCSVLEKVSKFEQREQGSQRPSVGGSGFGHNYRPHRTVSTSSTSGNDFEETKAHIRFSESAEP\nLGNGEQHFKNGELKLEEASRQPCGQQLSGGASDSGRGPQRPDARLLRSQSTFQLSSEPEREPEWRDRPGS\nPESPLLDAPFSRAYRNSIKDAQSRVLGATSFRRRDLELGAPVASRSWRPRPSSAHVGLRSPEASASASPH\nTPRERHSVTPAEGDLARPVPPAARRGARRRLTPEQKKRSYSEPEKMNEVGIVEEAEPAPLGPQRNGMRFP\nESSVADRRRLFERDGKACSTLSLSGPELKQFQQSALADYIQRKTGKRPTSAAGCSLQEPGPLRERAQSAY\nLQPGPAALEGSGLASASSLSSLREPSLQPRREATLLPATVAETQQAPRDRSSSFAGGRRLGERRRGDLLS\nGANGGTRGTQRGDETPREPSSWGARAGKSMSAEDLLERSDVLAGPVHVRSRSSPATADKRQDVLLGQDSG\nFGLVKDPCYLAGPGSRSLSCSERGQEEMLPLFHHLTPRWGGSGCKAIGDSSVPSECPGTLDHQRQASRTP\nCPRPPLAGTQGLVTDTRAAPLTPIGTPLPSAIPSGYCSQDGQTGRQPLPPYTPAMMHRSNGHTLTQPPGP\nRGCEGDGPEHGVEEGTRKRVSLPQWPPPSRAKWAHAAREDSLPEESSAPDFANLKHYQKQQSLPSLCSTS\nDPDTPLGAPSTPGRISLRISESVLRDSPPPHEDYEDEVFVRDPHPKATSSPTFEPLPPPPPPPPSQETPV\nYSMDDFPPPPPHTVCEAQLDSEDPEGPRPSFNKLSKVTIARERHMPGAAHVVGSQTLASRLQTSIKGSEA\nESTPPSFMSVHAQLAGSLGGQPAPIQTQSLSHDPVSGTQGLEKKVSPDPQKSSEDIRTEALAKEIVHQDK\nSLADILDPDSRLKTTMDLMEGLFPRDVNLLKENSVKRKAIQRTVSSSGCEGKRNEDKEAVSMLVNCPAYY\nSVSAPKAELLNKIKEMPAEVNEEEEQADVNEKKAELIGSLTHKLETLQEAKGSLLTDIKLNNALGEEVEA\nLISELCKPNEFDKYRMFIGDLDKVVNLLLSLSGRLARVENVLSGLGEDASNEERSSLYEKRKILAGQHED\nARELKENLDRRERVVLGILANYLSEEQLQDYQHFVKMKSTLLIEQRKLDDKIKLGQEQVKCLLESLPSDF\nIPKAGALALPPNLTSEPIPAGGCTFSGIFPTLTSPL\n\n&quot;</code></pre>
<p>We’ll use the <code>cat()</code> function to do a little formatting for us; it essentially just enforces the <strong>lines breaks</strong>:</p>
<div class="sourceCode" id="cb175"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb175-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb175-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(hShroom3)</span></code></pre></div>
<pre><code>## &gt;NP_065910.3 protein Shroom3 [Homo sapiens]
## MMRTTEDFHKPSATLNSNTATKGRYIYLEAFLEGGAPWGFTLKGGLEHGEPLIISKVEEGGKADTLSSKL
## QAGDEVVHINEVTLSSSRKEAVSLVKGSYKTLRLVVRRDVCTDPGHADTGASNFVSPEHLTSGPQHRKAA
## WSGGVKLRLKHRRSEPAGRPHSWHTTKSGEKQPDASMMQISQGMIGPPWHQSYHSSSSTSDLSNYDHAYL
## RRSPDQCSSQGSMESLEPSGAYPPCHLSPAKSTGSIDQLSHFHNKRDSAYSSFSTSSSILEYPHPGISGR
## ERSGSMDNTSARGGLLEGMRQADIRYVKTVYDTRRGVSAEYEVNSSALLLQGREARASANGQGYDKWSNI
## PRGKGVPPPSWSQQCPSSLETATDNLPPKVGAPLPPARSDSYAAFRHRERPSSWSSLDQKRLCRPQANSL
## GSLKSPFIEEQLHTVLEKSPENSPPVKPKHNYTQKAQPGQPLLPTSIYPVPSLEPHFAQVPQPSVSSNGM
## LYPALAKESGYIAPQGACNKMATIDENGNQNGSGRPGFAFCQPLEHDLLSPVEKKPEATAKYVPSKVHFC
## SVPENEEDASLKRHLTPPQGNSPHSNERKSTHSNKPSSHPHSLKCPQAQAWQAGEDKRSSRLSEPWEGDF
## QEDHNANLWRRLEREGLGQSLSGNFGKTKSAFSSLQNIPESLRRHSSLELGRGTQEGYPGGRPTCAVNTK
## AEDPGRKAAPDLGSHLDRQVSYPRPEGRTGASASFNSTDPSPEEPPAPSHPHTSSLGRRGPGPGSASALQ
## GFQYGKPHCSVLEKVSKFEQREQGSQRPSVGGSGFGHNYRPHRTVSTSSTSGNDFEETKAHIRFSESAEP
## LGNGEQHFKNGELKLEEASRQPCGQQLSGGASDSGRGPQRPDARLLRSQSTFQLSSEPEREPEWRDRPGS
## PESPLLDAPFSRAYRNSIKDAQSRVLGATSFRRRDLELGAPVASRSWRPRPSSAHVGLRSPEASASASPH
## TPRERHSVTPAEGDLARPVPPAARRGARRRLTPEQKKRSYSEPEKMNEVGIVEEAEPAPLGPQRNGMRFP
## ESSVADRRRLFERDGKACSTLSLSGPELKQFQQSALADYIQRKTGKRPTSAAGCSLQEPGPLRERAQSAY
## LQPGPAALEGSGLASASSLSSLREPSLQPRREATLLPATVAETQQAPRDRSSSFAGGRRLGERRRGDLLS
## GANGGTRGTQRGDETPREPSSWGARAGKSMSAEDLLERSDVLAGPVHVRSRSSPATADKRQDVLLGQDSG
## FGLVKDPCYLAGPGSRSLSCSERGQEEMLPLFHHLTPRWGGSGCKAIGDSSVPSECPGTLDHQRQASRTP
## CPRPPLAGTQGLVTDTRAAPLTPIGTPLPSAIPSGYCSQDGQTGRQPLPPYTPAMMHRSNGHTLTQPPGP
## RGCEGDGPEHGVEEGTRKRVSLPQWPPPSRAKWAHAAREDSLPEESSAPDFANLKHYQKQQSLPSLCSTS
## DPDTPLGAPSTPGRISLRISESVLRDSPPPHEDYEDEVFVRDPHPKATSSPTFEPLPPPPPPPPSQETPV
## YSMDDFPPPPPHTVCEAQLDSEDPEGPRPSFNKLSKVTIARERHMPGAAHVVGSQTLASRLQTSIKGSEA
## ESTPPSFMSVHAQLAGSLGGQPAPIQTQSLSHDPVSGTQGLEKKVSPDPQKSSEDIRTEALAKEIVHQDK
## SLADILDPDSRLKTTMDLMEGLFPRDVNLLKENSVKRKAIQRTVSSSGCEGKRNEDKEAVSMLVNCPAYY
## SVSAPKAELLNKIKEMPAEVNEEEEQADVNEKKAELIGSLTHKLETLQEAKGSLLTDIKLNNALGEEVEA
## LISELCKPNEFDKYRMFIGDLDKVVNLLLSLSGRLARVENVLSGLGEDASNEERSSLYEKRKILAGQHED
## ARELKENLDRRERVVLGILANYLSEEQLQDYQHFVKMKSTLLIEQRKLDDKIKLGQEQVKCLLESLPSDF
## IPKAGALALPPNLTSEPIPAGGCTFSGIFPTLTSPL</code></pre>
<p>Note the initial <code>&gt;</code>, then the header line of <code>NP_065910.3 protein Shroom3 [Homo sapiens]</code>. After that is the amino acid sequence. The underlying data also includes the <strong>newline character</strong> <code>\n</code> to designate where each line of amino acids stops (that is, the location of line breaks).</p>
<p>We can get the rest of the data by just changing the <code>id = ...</code> argument:</p>
<div class="sourceCode" id="cb177"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb177-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb177-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Mouse shroom 3a (M. musculus)</span></span>
<span id="cb177-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb177-2" aria-hidden="true" tabindex="-1"></a>mShroom3a <span class="ot">&lt;-</span> <span class="fu">entrez_fetch</span>(<span class="at">db =</span> <span class="st">&quot;protein&quot;</span>, </span>
<span id="cb177-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb177-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">id =</span> <span class="st">&quot;AAF13269&quot;</span>, </span>
<span id="cb177-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb177-4" aria-hidden="true" tabindex="-1"></a>                          <span class="at">rettype =</span> <span class="st">&quot;fasta&quot;</span>)</span>
<span id="cb177-5"><a href="worked-example-building-a-phylogeny-in-r.html#cb177-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb177-6"><a href="worked-example-building-a-phylogeny-in-r.html#cb177-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Human shroom 2 (H. sapiens)</span></span>
<span id="cb177-7"><a href="worked-example-building-a-phylogeny-in-r.html#cb177-7" aria-hidden="true" tabindex="-1"></a>hShroom2 <span class="ot">&lt;-</span> <span class="fu">entrez_fetch</span>(<span class="at">db =</span> <span class="st">&quot;protein&quot;</span>, </span>
<span id="cb177-8"><a href="worked-example-building-a-phylogeny-in-r.html#cb177-8" aria-hidden="true" tabindex="-1"></a>                          <span class="at">id =</span> <span class="st">&quot;CAA58534&quot;</span>, </span>
<span id="cb177-9"><a href="worked-example-building-a-phylogeny-in-r.html#cb177-9" aria-hidden="true" tabindex="-1"></a>                          <span class="at">rettype =</span> <span class="st">&quot;fasta&quot;</span>)</span>
<span id="cb177-10"><a href="worked-example-building-a-phylogeny-in-r.html#cb177-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb177-11"><a href="worked-example-building-a-phylogeny-in-r.html#cb177-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb177-12"><a href="worked-example-building-a-phylogeny-in-r.html#cb177-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Sea-urchin shroom</span></span>
<span id="cb177-13"><a href="worked-example-building-a-phylogeny-in-r.html#cb177-13" aria-hidden="true" tabindex="-1"></a>sShroom <span class="ot">&lt;-</span> <span class="fu">entrez_fetch</span>(<span class="at">db =</span> <span class="st">&quot;protein&quot;</span>, </span>
<span id="cb177-14"><a href="worked-example-building-a-phylogeny-in-r.html#cb177-14" aria-hidden="true" tabindex="-1"></a>                          <span class="at">id =</span> <span class="st">&quot;XP_783573&quot;</span>, </span>
<span id="cb177-15"><a href="worked-example-building-a-phylogeny-in-r.html#cb177-15" aria-hidden="true" tabindex="-1"></a>                          <span class="at">rettype =</span> <span class="st">&quot;fasta&quot;</span>)</span></code></pre></div>
<p>Here, I’ve pasted the function I used above three times into the code chunk and changed the id = … statement. Later in this script will avoid this clunky type of coding by using <strong>for loops</strong>.</p>
<p>I’m going to check about how long each of these sequences is - each should have an at least slightly different length. If any are identical, I might have repeated an accession name or re-used an object name. The function <code>nchar()</code> counts of the number of characters in an <em>R</em> object.</p>
<div class="sourceCode" id="cb178"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb178-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb178-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nchar</span>(hShroom3)</span></code></pre></div>
<pre><code>## [1] 2070</code></pre>
<div class="sourceCode" id="cb180"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb180-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb180-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nchar</span>(mShroom3a)</span></code></pre></div>
<pre><code>## [1] 2083</code></pre>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb182-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nchar</span>(sShroom)</span></code></pre></div>
<pre><code>## [1] 1758</code></pre>
<div class="sourceCode" id="cb184"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb184-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb184-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nchar</span>(hShroom2)</span></code></pre></div>
<pre><code>## [1] 1673</code></pre>
</div>
<div id="prepping-macromolecular-sequences" class="section level2" number="18.4">
<h2><span class="header-section-number">18.4</span> Prepping macromolecular sequences</h2>
<blockquote>
<p>“90% of data analysis is data cleaning” (-Just about every data analyst and data scientist)</p>
</blockquote>
<p>We have our sequences, but the current format isn’t directly usable for us yet because there are several things that aren’t sequence information</p>
<ol style="list-style-type: decimal">
<li>metadata (the header)</li>
<li>page formatting information (the newline character)</li>
</ol>
<p>We can remove this non-sequence information using a function I wrote called <code>fasta_cleaner()</code>, which is in the <code>compbio4all</code> package. The function uses <strong>regular expressions</strong> to remove the info we don’t need.</p>
<p>If you had trouble downloading the compbio4all package function you can add fasta_cleaner() to your R session directly by running this code:</p>
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb186-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb186-1" aria-hidden="true" tabindex="-1"></a>fasta_cleaner <span class="ot">&lt;-</span> <span class="cf">function</span>(fasta_object, <span class="at">parse =</span> <span class="cn">TRUE</span>){</span>
<span id="cb186-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb186-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb186-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb186-3" aria-hidden="true" tabindex="-1"></a>  fasta_object <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">&quot;^(&gt;)(.*?)(</span><span class="sc">\\</span><span class="st">n)(.*)(</span><span class="sc">\\</span><span class="st">n</span><span class="sc">\\</span><span class="st">n)&quot;</span>,<span class="st">&quot;</span><span class="sc">\\</span><span class="st">4&quot;</span>,fasta_object)</span>
<span id="cb186-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb186-4" aria-hidden="true" tabindex="-1"></a>  fasta_object <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>, <span class="st">&quot;&quot;</span>, fasta_object)</span>
<span id="cb186-5"><a href="worked-example-building-a-phylogeny-in-r.html#cb186-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb186-6"><a href="worked-example-building-a-phylogeny-in-r.html#cb186-6" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(parse <span class="sc">==</span> <span class="cn">TRUE</span>){</span>
<span id="cb186-7"><a href="worked-example-building-a-phylogeny-in-r.html#cb186-7" aria-hidden="true" tabindex="-1"></a>    fasta_object <span class="ot">&lt;-</span> stringr<span class="sc">::</span><span class="fu">str_split</span>(fasta_object,</span>
<span id="cb186-8"><a href="worked-example-building-a-phylogeny-in-r.html#cb186-8" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">pattern =</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb186-9"><a href="worked-example-building-a-phylogeny-in-r.html#cb186-9" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">simplify =</span> <span class="cn">FALSE</span>)</span>
<span id="cb186-10"><a href="worked-example-building-a-phylogeny-in-r.html#cb186-10" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb186-11"><a href="worked-example-building-a-phylogeny-in-r.html#cb186-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb186-12"><a href="worked-example-building-a-phylogeny-in-r.html#cb186-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(fasta_object[[<span class="dv">1</span>]])</span>
<span id="cb186-13"><a href="worked-example-building-a-phylogeny-in-r.html#cb186-13" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>If we run the name of the command with out any quotation marks we can see the code:</p>
<div class="sourceCode" id="cb187"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb187-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb187-1" aria-hidden="true" tabindex="-1"></a>fasta_cleaner</span></code></pre></div>
<pre><code>## function(fasta_object, parse = TRUE){
## 
##   fasta_object &lt;- sub(&quot;^(&gt;)(.*?)(\\n)(.*)(\\n\\n)&quot;,&quot;\\4&quot;,fasta_object)
##   fasta_object &lt;- gsub(&quot;\n&quot;, &quot;&quot;, fasta_object)
## 
##   if(parse == TRUE){
##     fasta_object &lt;- stringr::str_split(fasta_object,
##                                        pattern = &quot;&quot;,
##                                        simplify = FALSE)
##   }
## 
##   return(fasta_object[[1]])
## }</code></pre>
<p>Now use the function to clean our sequences; we won’t worry about what <code>parse = ...</code> is for.</p>
<div class="sourceCode" id="cb189"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb189-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb189-1" aria-hidden="true" tabindex="-1"></a>hShroom3  <span class="ot">&lt;-</span> <span class="fu">fasta_cleaner</span>(hShroom3,  <span class="at">parse =</span> F)</span>
<span id="cb189-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb189-2" aria-hidden="true" tabindex="-1"></a>mShroom3a <span class="ot">&lt;-</span> <span class="fu">fasta_cleaner</span>(mShroom3a, <span class="at">parse =</span> F)</span>
<span id="cb189-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb189-3" aria-hidden="true" tabindex="-1"></a>hShroom2  <span class="ot">&lt;-</span> <span class="fu">fasta_cleaner</span>(hShroom2,  <span class="at">parse =</span> F)</span>
<span id="cb189-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb189-4" aria-hidden="true" tabindex="-1"></a>sShroom   <span class="ot">&lt;-</span> <span class="fu">fasta_cleaner</span>(sShroom,   <span class="at">parse =</span> F)</span></code></pre></div>
<p>Again, I want to do something four times, so I’ve repeated the same line of code four times with the necessary change. This gets the job done, but there are better ways to do this using for loops.</p>
<p>Now let’s take a peek at what our sequences look like:</p>
<div class="sourceCode" id="cb190"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb190-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb190-1" aria-hidden="true" tabindex="-1"></a>hShroom3</span></code></pre></div>
<pre><code>## [1] &quot;MMRTTEDFHKPSATLNSNTATKGRYIYLEAFLEGGAPWGFTLKGGLEHGEPLIISKVEEGGKADTLSSKLQAGDEVVHINEVTLSSSRKEAVSLVKGSYKTLRLVVRRDVCTDPGHADTGASNFVSPEHLTSGPQHRKAAWSGGVKLRLKHRRSEPAGRPHSWHTTKSGEKQPDASMMQISQGMIGPPWHQSYHSSSSTSDLSNYDHAYLRRSPDQCSSQGSMESLEPSGAYPPCHLSPAKSTGSIDQLSHFHNKRDSAYSSFSTSSSILEYPHPGISGRERSGSMDNTSARGGLLEGMRQADIRYVKTVYDTRRGVSAEYEVNSSALLLQGREARASANGQGYDKWSNIPRGKGVPPPSWSQQCPSSLETATDNLPPKVGAPLPPARSDSYAAFRHRERPSSWSSLDQKRLCRPQANSLGSLKSPFIEEQLHTVLEKSPENSPPVKPKHNYTQKAQPGQPLLPTSIYPVPSLEPHFAQVPQPSVSSNGMLYPALAKESGYIAPQGACNKMATIDENGNQNGSGRPGFAFCQPLEHDLLSPVEKKPEATAKYVPSKVHFCSVPENEEDASLKRHLTPPQGNSPHSNERKSTHSNKPSSHPHSLKCPQAQAWQAGEDKRSSRLSEPWEGDFQEDHNANLWRRLEREGLGQSLSGNFGKTKSAFSSLQNIPESLRRHSSLELGRGTQEGYPGGRPTCAVNTKAEDPGRKAAPDLGSHLDRQVSYPRPEGRTGASASFNSTDPSPEEPPAPSHPHTSSLGRRGPGPGSASALQGFQYGKPHCSVLEKVSKFEQREQGSQRPSVGGSGFGHNYRPHRTVSTSSTSGNDFEETKAHIRFSESAEPLGNGEQHFKNGELKLEEASRQPCGQQLSGGASDSGRGPQRPDARLLRSQSTFQLSSEPEREPEWRDRPGSPESPLLDAPFSRAYRNSIKDAQSRVLGATSFRRRDLELGAPVASRSWRPRPSSAHVGLRSPEASASASPHTPRERHSVTPAEGDLARPVPPAARRGARRRLTPEQKKRSYSEPEKMNEVGIVEEAEPAPLGPQRNGMRFPESSVADRRRLFERDGKACSTLSLSGPELKQFQQSALADYIQRKTGKRPTSAAGCSLQEPGPLRERAQSAYLQPGPAALEGSGLASASSLSSLREPSLQPRREATLLPATVAETQQAPRDRSSSFAGGRRLGERRRGDLLSGANGGTRGTQRGDETPREPSSWGARAGKSMSAEDLLERSDVLAGPVHVRSRSSPATADKRQDVLLGQDSGFGLVKDPCYLAGPGSRSLSCSERGQEEMLPLFHHLTPRWGGSGCKAIGDSSVPSECPGTLDHQRQASRTPCPRPPLAGTQGLVTDTRAAPLTPIGTPLPSAIPSGYCSQDGQTGRQPLPPYTPAMMHRSNGHTLTQPPGPRGCEGDGPEHGVEEGTRKRVSLPQWPPPSRAKWAHAAREDSLPEESSAPDFANLKHYQKQQSLPSLCSTSDPDTPLGAPSTPGRISLRISESVLRDSPPPHEDYEDEVFVRDPHPKATSSPTFEPLPPPPPPPPSQETPVYSMDDFPPPPPHTVCEAQLDSEDPEGPRPSFNKLSKVTIARERHMPGAAHVVGSQTLASRLQTSIKGSEAESTPPSFMSVHAQLAGSLGGQPAPIQTQSLSHDPVSGTQGLEKKVSPDPQKSSEDIRTEALAKEIVHQDKSLADILDPDSRLKTTMDLMEGLFPRDVNLLKENSVKRKAIQRTVSSSGCEGKRNEDKEAVSMLVNCPAYYSVSAPKAELLNKIKEMPAEVNEEEEQADVNEKKAELIGSLTHKLETLQEAKGSLLTDIKLNNALGEEVEALISELCKPNEFDKYRMFIGDLDKVVNLLLSLSGRLARVENVLSGLGEDASNEERSSLYEKRKILAGQHEDARELKENLDRRERVVLGILANYLSEEQLQDYQHFVKMKSTLLIEQRKLDDKIKLGQEQVKCLLESLPSDFIPKAGALALPPNLTSEPIPAGGCTFSGIFPTLTSPL&quot;</code></pre>
<p>THe header and slash-n newline characters are gone. The sequence is now ready for use by our R alignment functions.</p>
</div>
<div id="aligning-sequences" class="section level2" number="18.5">
<h2><span class="header-section-number">18.5</span> Aligning sequences</h2>
<p>We can do a <a href="https://tinyurl.com/y4du73zq"><strong>global alignment</strong></a> of one sequence against another using the <code>pairwiseAlignment()</code> function from the <strong>Bioconductor</strong> package <code>Biostrings</code> (note that capital “B” in <code>Biostrings</code>; most <em>R</em> package names are all lower case, but not this one). Global alignment algorithms identify the best way to line up to sequences so that that optimal number of bases or amino acids are the same and the number of <strong>indels</strong> (insertions/deletions) are minimized. (Global alignment contrasts with <strong>local alignment</strong>, which works with portions of sequences and is used in database search programs like <strong>BLAST</strong>, the Basic Local Alignment Search Tool used by many biologists).</p>
<p>Let’s align human versus mouse shroom using the global alignment function pairwiseAlignment():</p>
<div class="sourceCode" id="cb192"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb192-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb192-1" aria-hidden="true" tabindex="-1"></a>align.h3.vs.m3a <span class="ot">&lt;-</span> Biostrings<span class="sc">::</span><span class="fu">pairwiseAlignment</span>(</span>
<span id="cb192-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb192-2" aria-hidden="true" tabindex="-1"></a>                  hShroom3,</span>
<span id="cb192-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb192-3" aria-hidden="true" tabindex="-1"></a>                  mShroom3a)</span></code></pre></div>
<p>We can peek at the alignment</p>
<div class="sourceCode" id="cb193"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb193-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb193-1" aria-hidden="true" tabindex="-1"></a>align.h3.vs.m3a</span></code></pre></div>
<pre><code>## Global PairwiseAlignmentsSingleSubject (1 of 1)
## pattern: MMRTTEDFHKPSATLN-SNTATKGRYIYLEAFLE...KAGALALPPNLTSEPIPAGGCTFSGIFPTLTSPL
## subject: MK-TPENLEEPSATPNPSRTPTE-RFVYLEALLE...KAGAISLPPALTGHATPGGTSVFGGVFPTLTSPL
## score: 2189.934</code></pre>
<p>The <strong>score</strong> tells us how closely they are aligned; higher scores mean the sequences are more similar. In general, perfect matches increase scores the most, and indels decrease scores.</p>
<p>Its hard to interpret scores on their own so we can get the <strong>percent sequence identity (PID)</strong> (aka percent identical, proportion identity, proportion identical) using the <code>pid()</code> function.</p>
<div class="sourceCode" id="cb195"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb195-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb195-1" aria-hidden="true" tabindex="-1"></a>Biostrings<span class="sc">::</span><span class="fu">pid</span>(align.h3.vs.m3a)</span></code></pre></div>
<pre><code>## [1] 70.56511</code></pre>
<p>So, <em>shroom3</em> from humans (hShroom3) and <em>shroom3</em> from mice (mShroom3) are ~71% similar (at least using this particular method of alignment, and there are many ways to do this!)</p>
<p>What about human shroom 3 and human shroom 2? Shroom is a <strong>gene family</strong>, and there are different versions of the gene within a genome.</p>
<div class="sourceCode" id="cb197"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb197-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb197-1" aria-hidden="true" tabindex="-1"></a>align.h3.vs.h2 <span class="ot">&lt;-</span> Biostrings<span class="sc">::</span><span class="fu">pairwiseAlignment</span>(</span>
<span id="cb197-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb197-2" aria-hidden="true" tabindex="-1"></a>                  hShroom3,</span>
<span id="cb197-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb197-3" aria-hidden="true" tabindex="-1"></a>                  hShroom2)</span></code></pre></div>
<p>If you take a look at the alignment you can see there are a lot of indels</p>
<div class="sourceCode" id="cb198"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb198-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb198-1" aria-hidden="true" tabindex="-1"></a>align.h3.vs.h2</span></code></pre></div>
<pre><code>## Global PairwiseAlignmentsSingleSubject (1 of 1)
## pattern: MMRTTEDFHKPSATLNSNT--ATKGRYIYLEAFL...KAGALALPPNLTSEPIPAGGCTFSGIFPTLTSPL
## subject: MEGA-EPRARPERLAEAETRAADGGRLV--EVQL...----------------PERGK-------------
## score: -5673.853</code></pre>
<p>Check out the score itself using <code>score()</code>, which accesses it directly without all the other information.</p>
<div class="sourceCode" id="cb200"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb200-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb200-1" aria-hidden="true" tabindex="-1"></a><span class="fu">score</span>(align.h3.vs.h2)</span></code></pre></div>
<pre><code>## [1] -5673.853</code></pre>
<p>Its negative because there are a LOT of indels.</p>
<p>Now the percent sequence alignment with <code>pid()</code>:</p>
<div class="sourceCode" id="cb202"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb202-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb202-1" aria-hidden="true" tabindex="-1"></a>Biostrings<span class="sc">::</span><span class="fu">pid</span>(align.h3.vs.h2)</span></code></pre></div>
<pre><code>## [1] 33.83277</code></pre>
<p>So Human shroom 3 and Mouse shroom 3 are 71% identical, but Human shroom 3 and human shroom 2 are only 34% similar? How does it work out evolutionarily that a human and mouse gene are more similar than a human and a human gene? What are the evolutionary relationships among these genes within the shroom gene family?</p>
<p>An important note: indels contribute negatively to an alignment score, but aren’t used in the most common calculations for PID.</p>
</div>
<div id="the-shroom-family-of-genes" class="section level2" number="18.6">
<h2><span class="header-section-number">18.6</span> The shroom family of genes</h2>
<p>I’ve copied a table from a published paper which has accession numbers for 15 different Shroom genes. Some are different types of shroom from the same organism (e.g. shroom 1, 2, 3 and 4 from humans), or they are from different organism (e.g. frogs, mice, bees).</p>
<div class="sourceCode" id="cb204"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb204-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb204-1" aria-hidden="true" tabindex="-1"></a>shroom_table <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;CAA78718&quot;</span> , <span class="st">&quot;X. laevis Apx&quot;</span> ,         <span class="st">&quot;xShroom1&quot;</span>,</span>
<span id="cb204-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb204-2" aria-hidden="true" tabindex="-1"></a>                  <span class="st">&quot;NP_597713&quot;</span> , <span class="st">&quot;H. sapiens APXL2&quot;</span> ,     <span class="st">&quot;hShroom1&quot;</span>,</span>
<span id="cb204-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb204-3" aria-hidden="true" tabindex="-1"></a>                  <span class="st">&quot;CAA58534&quot;</span> , <span class="st">&quot;H. sapiens APXL&quot;</span>,        <span class="st">&quot;hShroom2&quot;</span>,</span>
<span id="cb204-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb204-4" aria-hidden="true" tabindex="-1"></a>                  <span class="st">&quot;ABD19518&quot;</span> , <span class="st">&quot;M. musculus Apxl&quot;</span> ,      <span class="st">&quot;mShroom2&quot;</span>,</span>
<span id="cb204-5"><a href="worked-example-building-a-phylogeny-in-r.html#cb204-5" aria-hidden="true" tabindex="-1"></a>                  <span class="st">&quot;AAF13269&quot;</span> , <span class="st">&quot;M. musculus ShroomL&quot;</span> ,   <span class="st">&quot;mShroom3a&quot;</span>,</span>
<span id="cb204-6"><a href="worked-example-building-a-phylogeny-in-r.html#cb204-6" aria-hidden="true" tabindex="-1"></a>                  <span class="st">&quot;AAF13270&quot;</span> , <span class="st">&quot;M. musculus ShroomS&quot;</span> ,   <span class="st">&quot;mShroom3b&quot;</span>,</span>
<span id="cb204-7"><a href="worked-example-building-a-phylogeny-in-r.html#cb204-7" aria-hidden="true" tabindex="-1"></a>                  <span class="st">&quot;NP_065910&quot;</span>, <span class="st">&quot;H. sapiens Shroom&quot;</span> ,     <span class="st">&quot;hShroom3&quot;</span>,</span>
<span id="cb204-8"><a href="worked-example-building-a-phylogeny-in-r.html#cb204-8" aria-hidden="true" tabindex="-1"></a>                  <span class="st">&quot;ABD59319&quot;</span> , <span class="st">&quot;X. laevis Shroom-like&quot;</span>,  <span class="st">&quot;xShroom3&quot;</span>,</span>
<span id="cb204-9"><a href="worked-example-building-a-phylogeny-in-r.html#cb204-9" aria-hidden="true" tabindex="-1"></a>                  <span class="st">&quot;NP_065768&quot;</span>, <span class="st">&quot;H. sapiens KIAA1202&quot;</span> ,   <span class="st">&quot;hShroom4a&quot;</span>,</span>
<span id="cb204-10"><a href="worked-example-building-a-phylogeny-in-r.html#cb204-10" aria-hidden="true" tabindex="-1"></a>                  <span class="st">&quot;AAK95579&quot;</span> , <span class="st">&quot;H. sapiens SHAP-A&quot;</span> ,     <span class="st">&quot;hShroom4b&quot;</span>,</span>
<span id="cb204-11"><a href="worked-example-building-a-phylogeny-in-r.html#cb204-11" aria-hidden="true" tabindex="-1"></a>                 <span class="co">#&quot;DQ435686&quot; , &quot;M. musculus KIAA1202&quot; ,  &quot;mShroom4&quot;,</span></span>
<span id="cb204-12"><a href="worked-example-building-a-phylogeny-in-r.html#cb204-12" aria-hidden="true" tabindex="-1"></a>                  <span class="st">&quot;ABA81834&quot;</span> , <span class="st">&quot;D. melanogaster Shroom&quot;</span>, <span class="st">&quot;dmShroom&quot;</span>,</span>
<span id="cb204-13"><a href="worked-example-building-a-phylogeny-in-r.html#cb204-13" aria-hidden="true" tabindex="-1"></a>                  <span class="st">&quot;EAA12598&quot;</span> , <span class="st">&quot;A. gambiae Shroom&quot;</span>,      <span class="st">&quot;agShroom&quot;</span>,</span>
<span id="cb204-14"><a href="worked-example-building-a-phylogeny-in-r.html#cb204-14" aria-hidden="true" tabindex="-1"></a>                  <span class="st">&quot;XP_392427&quot;</span> , <span class="st">&quot;A. mellifera Shroom&quot;</span> ,  <span class="st">&quot;amShroom&quot;</span>,</span>
<span id="cb204-15"><a href="worked-example-building-a-phylogeny-in-r.html#cb204-15" aria-hidden="true" tabindex="-1"></a>                  <span class="st">&quot;XP_783573&quot;</span> , <span class="st">&quot;S. purpuratus Shroom&quot;</span> , <span class="st">&quot;spShroom&quot;</span>) <span class="co">#sea urchin</span></span></code></pre></div>
<p>What we just made is just one long vector with all the info.</p>
<div class="sourceCode" id="cb205"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb205-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb205-1" aria-hidden="true" tabindex="-1"></a><span class="fu">is</span>(shroom_table)</span></code></pre></div>
<pre><code>##  [1] &quot;character&quot;               &quot;vector&quot;                 
##  [3] &quot;data.frameRowLabels&quot;     &quot;SuperClassMethod&quot;       
##  [5] &quot;EnumerationValue&quot;        &quot;character_OR_connection&quot;
##  [7] &quot;character_OR_NULL&quot;       &quot;atomic&quot;                 
##  [9] &quot;vector_OR_Vector&quot;        &quot;vector_OR_factor&quot;</code></pre>
<div class="sourceCode" id="cb207"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb207-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb207-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(shroom_table)</span></code></pre></div>
<pre><code>## [1] &quot;character&quot;</code></pre>
<div class="sourceCode" id="cb209"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb209-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb209-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(shroom_table)</span></code></pre></div>
<pre><code>## [1] 42</code></pre>
<p>I’ll do a bit of formatting; you can ignore these details if you want</p>
<div class="sourceCode" id="cb211"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb211-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb211-1" aria-hidden="true" tabindex="-1"></a><span class="co"># convert the vector to matrix using matrix()</span></span>
<span id="cb211-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb211-2" aria-hidden="true" tabindex="-1"></a>shroom_table_matrix <span class="ot">&lt;-</span> <span class="fu">matrix</span>(shroom_table,</span>
<span id="cb211-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb211-3" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">byrow =</span> T,</span>
<span id="cb211-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb211-4" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">nrow =</span> <span class="dv">14</span>)</span>
<span id="cb211-5"><a href="worked-example-building-a-phylogeny-in-r.html#cb211-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb211-6"><a href="worked-example-building-a-phylogeny-in-r.html#cb211-6" aria-hidden="true" tabindex="-1"></a><span class="co"># convert the matrix to a dataframe using data.frame()</span></span>
<span id="cb211-7"><a href="worked-example-building-a-phylogeny-in-r.html#cb211-7" aria-hidden="true" tabindex="-1"></a>shroom_table <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(shroom_table_matrix, </span>
<span id="cb211-8"><a href="worked-example-building-a-phylogeny-in-r.html#cb211-8" aria-hidden="true" tabindex="-1"></a>                     <span class="at">stringsAsFactors =</span> F)</span>
<span id="cb211-9"><a href="worked-example-building-a-phylogeny-in-r.html#cb211-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb211-10"><a href="worked-example-building-a-phylogeny-in-r.html#cb211-10" aria-hidden="true" tabindex="-1"></a><span class="co"># name columns of dataframe using names() function</span></span>
<span id="cb211-11"><a href="worked-example-building-a-phylogeny-in-r.html#cb211-11" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(shroom_table) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;accession&quot;</span>, <span class="st">&quot;name.orig&quot;</span>,<span class="st">&quot;name.new&quot;</span>)</span>
<span id="cb211-12"><a href="worked-example-building-a-phylogeny-in-r.html#cb211-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb211-13"><a href="worked-example-building-a-phylogeny-in-r.html#cb211-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Create simplified species names</span></span>
<span id="cb211-14"><a href="worked-example-building-a-phylogeny-in-r.html#cb211-14" aria-hidden="true" tabindex="-1"></a><span class="do">## access species column using $ notation</span></span>
<span id="cb211-15"><a href="worked-example-building-a-phylogeny-in-r.html#cb211-15" aria-hidden="true" tabindex="-1"></a>shroom_table<span class="sc">$</span>spp <span class="ot">&lt;-</span> <span class="st">&quot;Homo&quot;</span></span>
<span id="cb211-16"><a href="worked-example-building-a-phylogeny-in-r.html#cb211-16" aria-hidden="true" tabindex="-1"></a>shroom_table<span class="sc">$</span>spp[<span class="fu">grep</span>(<span class="st">&quot;laevis&quot;</span>,shroom_table<span class="sc">$</span>name.orig)] <span class="ot">&lt;-</span> <span class="st">&quot;Xenopus&quot;</span></span>
<span id="cb211-17"><a href="worked-example-building-a-phylogeny-in-r.html#cb211-17" aria-hidden="true" tabindex="-1"></a>shroom_table<span class="sc">$</span>spp[<span class="fu">grep</span>(<span class="st">&quot;musculus&quot;</span>,shroom_table<span class="sc">$</span>name.orig)] <span class="ot">&lt;-</span> <span class="st">&quot;Mus&quot;</span></span>
<span id="cb211-18"><a href="worked-example-building-a-phylogeny-in-r.html#cb211-18" aria-hidden="true" tabindex="-1"></a>shroom_table<span class="sc">$</span>spp[<span class="fu">grep</span>(<span class="st">&quot;melanogaster&quot;</span>,shroom_table<span class="sc">$</span>name.orig)] <span class="ot">&lt;-</span> <span class="st">&quot;Drosophila&quot;</span></span>
<span id="cb211-19"><a href="worked-example-building-a-phylogeny-in-r.html#cb211-19" aria-hidden="true" tabindex="-1"></a>shroom_table<span class="sc">$</span>spp[<span class="fu">grep</span>(<span class="st">&quot;gambiae&quot;</span>,shroom_table<span class="sc">$</span>name.orig)] <span class="ot">&lt;-</span> <span class="st">&quot;mosquito&quot;</span></span>
<span id="cb211-20"><a href="worked-example-building-a-phylogeny-in-r.html#cb211-20" aria-hidden="true" tabindex="-1"></a>shroom_table<span class="sc">$</span>spp[<span class="fu">grep</span>(<span class="st">&quot;mellifera&quot;</span>,shroom_table<span class="sc">$</span>name.orig)] <span class="ot">&lt;-</span> <span class="st">&quot;bee&quot;</span></span>
<span id="cb211-21"><a href="worked-example-building-a-phylogeny-in-r.html#cb211-21" aria-hidden="true" tabindex="-1"></a>shroom_table<span class="sc">$</span>spp[<span class="fu">grep</span>(<span class="st">&quot;purpuratus&quot;</span>,shroom_table<span class="sc">$</span>name.orig)] <span class="ot">&lt;-</span> <span class="st">&quot;sea urchin&quot;</span></span></code></pre></div>
<p>Take a look at the finished table</p>
<div class="sourceCode" id="cb212"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb212-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb212-1" aria-hidden="true" tabindex="-1"></a>shroom_table</span></code></pre></div>
<pre><code>##    accession              name.orig  name.new        spp
## 1   CAA78718          X. laevis Apx  xShroom1    Xenopus
## 2  NP_597713       H. sapiens APXL2  hShroom1       Homo
## 3   CAA58534        H. sapiens APXL  hShroom2       Homo
## 4   ABD19518       M. musculus Apxl  mShroom2        Mus
## 5   AAF13269    M. musculus ShroomL mShroom3a        Mus
## 6   AAF13270    M. musculus ShroomS mShroom3b        Mus
## 7  NP_065910      H. sapiens Shroom  hShroom3       Homo
## 8   ABD59319  X. laevis Shroom-like  xShroom3    Xenopus
## 9  NP_065768    H. sapiens KIAA1202 hShroom4a       Homo
## 10  AAK95579      H. sapiens SHAP-A hShroom4b       Homo
## 11  ABA81834 D. melanogaster Shroom  dmShroom Drosophila
## 12  EAA12598      A. gambiae Shroom  agShroom   mosquito
## 13 XP_392427    A. mellifera Shroom  amShroom        bee
## 14 XP_783573   S. purpuratus Shroom  spShroom sea urchin</code></pre>
</div>
<div id="downloading-multiple-sequences" class="section level2" number="18.7">
<h2><span class="header-section-number">18.7</span> Downloading multiple sequences</h2>
<p>Instead of getting one sequence at a time we can download several by accessing the “accession” column from the table</p>
<pre class="{r}?"><code>shroom_table$accession</code></pre>
<p>We can give this whole set of accessions to <code>entrez_fetch()</code>, which is a <strong>vectorized function</strong> which knows how to handle a vector of inputs.</p>
<div class="sourceCode" id="cb215"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb215-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb215-1" aria-hidden="true" tabindex="-1"></a>shrooms <span class="ot">&lt;-</span> <span class="fu">entrez_fetch</span>(<span class="at">db =</span> <span class="st">&quot;protein&quot;</span>, </span>
<span id="cb215-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb215-2" aria-hidden="true" tabindex="-1"></a>                          <span class="at">id =</span> shroom_table<span class="sc">$</span>accession, </span>
<span id="cb215-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb215-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">rettype =</span> <span class="st">&quot;fasta&quot;</span>)</span></code></pre></div>
<p>We can look at what we got here with <code>cat()</code> (I won’t display this because it is very long!)</p>
<div class="sourceCode" id="cb216"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb216-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb216-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(shrooms)</span></code></pre></div>
<p>The current format of these data is a single, very long set of data. This is a standard way to store, share and transmit FASTA files, but in <em>R</em> we’ll need a slightly different format.</p>
<p>We’ll download all of the sequences again, this time using a function from <code>compbio4all</code> called <code>entrez_fetch_list()</code> which is a <strong>wrapper</strong> function I wrote to put the output of <code>entrez_fetch()</code> into an <em>R</em> data format called a <strong>list</strong>.</p>
<p>This function is contained with the combio4all package; however, if you are having trouble with the package you can enter it directly into your R sessions by running the source code of the function.</p>
<div class="sourceCode" id="cb217"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb217-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb217-1" aria-hidden="true" tabindex="-1"></a>entrez_fetch_list <span class="ot">&lt;-</span> <span class="cf">function</span>(db, id, rettype, ...){</span>
<span id="cb217-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb217-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb217-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb217-3" aria-hidden="true" tabindex="-1"></a>  <span class="co">#setup list for storing output</span></span>
<span id="cb217-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb217-4" aria-hidden="true" tabindex="-1"></a>  n.seq <span class="ot">&lt;-</span> <span class="fu">length</span>(id)</span>
<span id="cb217-5"><a href="worked-example-building-a-phylogeny-in-r.html#cb217-5" aria-hidden="true" tabindex="-1"></a>  list.output <span class="ot">&lt;-</span> <span class="fu">as.list</span>(<span class="fu">rep</span>(<span class="cn">NA</span>, n.seq))</span>
<span id="cb217-6"><a href="worked-example-building-a-phylogeny-in-r.html#cb217-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">names</span>(list.output) <span class="ot">&lt;-</span> id</span>
<span id="cb217-7"><a href="worked-example-building-a-phylogeny-in-r.html#cb217-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb217-8"><a href="worked-example-building-a-phylogeny-in-r.html#cb217-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># get output</span></span>
<span id="cb217-9"><a href="worked-example-building-a-phylogeny-in-r.html#cb217-9" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(id)){</span>
<span id="cb217-10"><a href="worked-example-building-a-phylogeny-in-r.html#cb217-10" aria-hidden="true" tabindex="-1"></a>    list.output[[i]] <span class="ot">&lt;-</span> rentrez<span class="sc">::</span><span class="fu">entrez_fetch</span>(<span class="at">db =</span> db,</span>
<span id="cb217-11"><a href="worked-example-building-a-phylogeny-in-r.html#cb217-11" aria-hidden="true" tabindex="-1"></a>                                              <span class="at">id =</span> id[i],</span>
<span id="cb217-12"><a href="worked-example-building-a-phylogeny-in-r.html#cb217-12" aria-hidden="true" tabindex="-1"></a>                                              <span class="at">rettype =</span> rettype)</span>
<span id="cb217-13"><a href="worked-example-building-a-phylogeny-in-r.html#cb217-13" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb217-14"><a href="worked-example-building-a-phylogeny-in-r.html#cb217-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb217-15"><a href="worked-example-building-a-phylogeny-in-r.html#cb217-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb217-16"><a href="worked-example-building-a-phylogeny-in-r.html#cb217-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(list.output)</span>
<span id="cb217-17"><a href="worked-example-building-a-phylogeny-in-r.html#cb217-17" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>However we get the function, we can use it to download a bunch of FASTA files and store them in an R list.</p>
<div class="sourceCode" id="cb218"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb218-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb218-1" aria-hidden="true" tabindex="-1"></a>shrooms_list <span class="ot">&lt;-</span> <span class="fu">entrez_fetch_list</span>(<span class="at">db =</span> <span class="st">&quot;protein&quot;</span>, </span>
<span id="cb218-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb218-2" aria-hidden="true" tabindex="-1"></a>                          <span class="at">id =</span> shroom_table<span class="sc">$</span>accession, </span>
<span id="cb218-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb218-3" aria-hidden="true" tabindex="-1"></a>                          <span class="at">rettype =</span> <span class="st">&quot;fasta&quot;</span>)</span></code></pre></div>
<p>Now we have an R <strong>list</strong> which as 14 <strong>elements</strong>, one for each sequence in our table.</p>
<div class="sourceCode" id="cb219"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb219-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb219-1" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(shrooms_list)</span></code></pre></div>
<pre><code>## [1] 14</code></pre>
<p>Each element of the list contains a FASTA entry for one sequence</p>
<div class="sourceCode" id="cb221"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb221-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb221-1" aria-hidden="true" tabindex="-1"></a>shrooms_list[[<span class="dv">1</span>]]</span></code></pre></div>
<pre><code>## [1] &quot;&gt;CAA78718.1 apical protein [Xenopus laevis]\nMSAFGNTIERWNIKSTGVIAGLGHSERISPVRSMTTLVDSAYSSFSGSSYVPEYQNSFQHDGCHYNDEQL\nSYMDSEYVRAIYNPSLLDKDGVYNDIVSEHGSSKVALSGRSSSSLCSDNTTSVHRTSPAKLDNYVTNLDS\nEKNIYGDPINMKHKQNRPNHKAYGLQRNSPTGINSLQEKENQLYNPSNFMEIKDNYFGRSLDVLQADGDI\nMTQDSYTQNALYFPQNQPDQYRNTQYPGANRMSKEQFKVNDVQKSNEENTERDGPYLTKDGQFVQGQYAS\nDVRTSFKNIRRSLKKSASGKIVAHDSQGSCWIMKPGKDTPSFNSEGTITDMDYDNREQWDIRKSRLSTRA\nSQSLYYESNEDVSGPPLKAMNSKNEVDQTLSFQKDATVKSIPLLSQQLQQEKCKSHPLSDLNCEKITKAS\nTPMLYHLAGGRHSAFIAPVHNTNPAQQEKLKLESKTLERMNNISVLQLSEPRPDNHKLPKNKSLTQLADL\nHDSVEGGNSGNLNSSAEESLMNDYIEKLKVAQKKVLRETSFKRKDLQMSLPCRFKLNPPKRPTIDHFRSY\nSSSSANEESAYLQTKNSADSSYKKDDTEKVAVTRIGGRKRITKEQKKLCYSEPEKLDHLGIQKSNFAWKE\nEPTFANRREMSDSDISANRIKYLESKERTNSSSNLSKTELKQIQHNALVQYMERKTNQRPNSNPQVQMER\nTSLGLPNYNEWSIYSSETSSSDASQKYLRRRSAGASSSYDATVTWNDRFGKTSPLGRSAAEKTAGVQRKT\nFSDQRTLDGSQEHLEGSSPSLSQKTSKSTHNEQVSYVNMEFLPSSHSKNHMYNDRLTVPGDGTSAESGRM\nFVSKSRGKSMEEIGTTDIVKLAELSHSSDQLYHIKGPVISSRLENTRTTAASHQDRLLASTQIETGNLPR\nQTHQESVVGPCRSDLANLGQEAHSWPLRASDVSPGTDNPCSSSPSAEVQPGAPEPLHCLQTEDEVFTPAS\nTARNEEPNSTAFSYLLSTGKPVSQGEATALSFTFLPEQDRLEHPIVSETTPSSESDENVSDAAAEKETTT\nTQLPETSNVNKPLGFTVDNQEVEGDGEPMQPEFIDSSKQLELSSLPSSQVNIMQTAEPYLGDKNIGNEQK\nTEDLEQKSKNPEEDDLPKVKLKSPEDEILEELVKEIVAKDKSLLNCLQPVSVRESAMDLMKSLFPMDVTA\nAEKSRTRGLLGKDKGETLKKNNSDLESSSKLPSKITGMLQKRPDGESLDDITLKKMELLSKIGSKLEDLC\nEQREFLLSDISKNTTNGNNMQTMVKELCKPNEFERYMMFIGDLEKVVSLLFSLSTRLTRVENSLSKVDEN\nTDAEEMQSLKERHNLLSSQREDAKDLKANLDRREQVVTGILVKYLNEEQLQDYKHFVRLKTSLLIEQKNL\nEEKIKVYEEQFESIHNSLPP\n\n&quot;</code></pre>
<p>We now need to clean up each one of these sequences. We could do this by running our fasta_cleaner() function on each of the elements of the list like this.</p>
<div class="sourceCode" id="cb223"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb223-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb223-1" aria-hidden="true" tabindex="-1"></a><span class="co"># clean sequence 1 in element 1 of list</span></span>
<span id="cb223-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb223-2" aria-hidden="true" tabindex="-1"></a>shrooms_list[[<span class="dv">1</span>]] <span class="ot">&lt;-</span> <span class="fu">fasta_cleaner</span>(shrooms_list[[<span class="dv">1</span>]], <span class="at">parse =</span> F)</span>
<span id="cb223-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb223-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb223-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb223-4" aria-hidden="true" tabindex="-1"></a><span class="co"># clean sequence 2 in element 2 of list</span></span>
<span id="cb223-5"><a href="worked-example-building-a-phylogeny-in-r.html#cb223-5" aria-hidden="true" tabindex="-1"></a>shrooms_list[[<span class="dv">2</span>]] <span class="ot">&lt;-</span> <span class="fu">fasta_cleaner</span>(shrooms_list[[<span class="dv">2</span>]], <span class="at">parse =</span> F)</span>
<span id="cb223-6"><a href="worked-example-building-a-phylogeny-in-r.html#cb223-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb223-7"><a href="worked-example-building-a-phylogeny-in-r.html#cb223-7" aria-hidden="true" tabindex="-1"></a><span class="co"># clean sequence 3 in element 3 of list</span></span>
<span id="cb223-8"><a href="worked-example-building-a-phylogeny-in-r.html#cb223-8" aria-hidden="true" tabindex="-1"></a>shrooms_list[[<span class="dv">3</span>]] <span class="ot">&lt;-</span> <span class="fu">fasta_cleaner</span>(shrooms_list[[<span class="dv">3</span>]], <span class="at">parse =</span> F)</span>
<span id="cb223-9"><a href="worked-example-building-a-phylogeny-in-r.html#cb223-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb223-10"><a href="worked-example-building-a-phylogeny-in-r.html#cb223-10" aria-hidden="true" tabindex="-1"></a><span class="co"># clean sequence x in element x of list</span></span>
<span id="cb223-11"><a href="worked-example-building-a-phylogeny-in-r.html#cb223-11" aria-hidden="true" tabindex="-1"></a><span class="do">## ...</span></span></code></pre></div>
<p>Copying the same line of code 14 time and making the necessary changes takes time and is error prone. We can do this more easily using a simple <code>for()</code> loop:</p>
<div class="sourceCode" id="cb224"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb224-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb224-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(shrooms_list)){</span>
<span id="cb224-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb224-2" aria-hidden="true" tabindex="-1"></a>  shrooms_list[[i]] <span class="ot">&lt;-</span> <span class="fu">fasta_cleaner</span>(shrooms_list[[i]], <span class="at">parse =</span> F)</span>
<span id="cb224-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb224-3" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>For loops all start with for(…) and contain the code we want to do repeatedly within curly brackets {…}. We won’t worry about the details of how for() loops work in R, but the upshot is that they allow us to easily repeatedly the same step many times while having making the necessary minor changes.</p>
<p>Now, for the second to last step of data preparation: we need to take each one of our sequences from our list and put it into a <strong>vector</strong>, in particular a <strong>named vector</strong>.</p>
<p>First, we need a vector to store stuff. We’ll make an empty vector that just has NAs in it using the rep() function.</p>
<div class="sourceCode" id="cb225"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb225-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb225-1" aria-hidden="true" tabindex="-1"></a>shrooms_vector <span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA</span>, <span class="fu">length</span>(shrooms_list))</span></code></pre></div>
<p>The result looks like this</p>
<div class="sourceCode" id="cb226"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb226-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb226-1" aria-hidden="true" tabindex="-1"></a>shrooms_vector</span></code></pre></div>
<pre><code>##  [1] NA NA NA NA NA NA NA NA NA NA NA NA NA NA</code></pre>
<p>Now we use a for() loop to take each element of shrooms_list and put it into the vector shrooms_vector (the precise details don’t matter; what is important is that we are using a for loop so we don’t have to repeat the same line of code 14 times)</p>
<div class="sourceCode" id="cb228"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb228-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb228-1" aria-hidden="true" tabindex="-1"></a><span class="co"># run the loop</span></span>
<span id="cb228-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb228-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(shrooms_vector)){</span>
<span id="cb228-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb228-3" aria-hidden="true" tabindex="-1"></a>  shrooms_vector[i] <span class="ot">&lt;-</span> shrooms_list[[i]]</span>
<span id="cb228-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb228-4" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
<p>Now we name the vector. This is using the names() function. (The exact details don’t matter)</p>
<div class="sourceCode" id="cb229"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb229-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb229-1" aria-hidden="true" tabindex="-1"></a><span class="co"># name the vector</span></span>
<span id="cb229-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb229-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(shrooms_vector) <span class="ot">&lt;-</span> <span class="fu">names</span>(shrooms_list)</span></code></pre></div>
<p>Now the final step: we need to convert our named vector to a <strong>string set</strong> using <code>Biostrings::AAStringSet()</code>. Note the <code>_ss</code> tag at the end of the object we’re assigning the output to, which designates this as a string set.</p>
<p>A string set is just a type of data format someone came up with to organize and annotate sequence data. (All of these steps are annoying, and someday I’ll write some functions to simplify all of this).</p>
<div class="sourceCode" id="cb230"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb230-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb230-1" aria-hidden="true" tabindex="-1"></a>shrooms_vector_ss <span class="ot">&lt;-</span> Biostrings<span class="sc">::</span><span class="fu">AAStringSet</span>(shrooms_vector)</span></code></pre></div>
</div>
<div id="multiple-sequence-alignment" class="section level2" number="18.8">
<h2><span class="header-section-number">18.8</span> Multiple sequence alignment</h2>
<p>We must <strong>align</strong> all of the sequences we downloaded and use that <strong>alignment</strong> to build a <strong>phylogenetic tree</strong>. This will tell us how the different genes, both within and between species, are likely to be related.</p>
<p>(Note: previously we’ve explored the similarities between sequences using pairwiseAlignments(). This helped us understand the data, but wasn’t actually necessary for making an MSA or phylogenetic tree).</p>
<div id="building-an-multiple-sequence-alignment-msa" class="section level3" number="18.8.1">
<h3><span class="header-section-number">18.8.1</span> Building an Multiple Sequence Alignment (MSA)</h3>
<p>Multiple sequence alignments (MSA) appear frequently in papers in molecular biology, biochemistry, and molecular evolution. They are also the basis for almost all phylogenies made with modern software to build phylogenetic trees using macromolecues. MSA are extensions of global alignment of two sequences. However, while a function like pairwiseAlignment() tries to come up with the best way to align two sequences, and MSA algorithm tries to come up with the joint alignment that is best across all alignments. This not only takes longer to do, but can sometimes come up with slight different results than a bunch of individual pairs of alignments.</p>
<p>We’ll use the software <code>msa,</code> which implements the <strong>ClustalW</strong> multiple sequence alignment algorithm. Normally we’d have to download the ClustalW program and either point-and-click our way through it or use the <strong>command line</strong>*, but these folks wrote up the algorithm in R so we can do this with a line of R code. This will take a second or two.</p>
<p>NOTE: While based in R, the msa package uses an R package called Rcpp (“R C++”) to integrate R with code from the language C++. There seem to some issues related to this process on some computers. If you can’t get msa to load or msa() to fun, you can comment-out the msa-related code.</p>
<div class="sourceCode" id="cb231"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb231-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb231-1" aria-hidden="true" tabindex="-1"></a>shrooms_align <span class="ot">&lt;-</span> <span class="fu">msa</span>(shrooms_vector_ss,</span>
<span id="cb231-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb231-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">method =</span> <span class="st">&quot;ClustalW&quot;</span>)</span></code></pre></div>
<pre><code>## use default substitution matrix</code></pre>
<p>While msa() runs R tells you “use default substitution matrix”, which means its using the programs default way of scoring alignments; that is, how to assign values to matches, mismatches, and indels while trying to come up with the best alignment of all the sequences.</p>
</div>
<div id="viewing-an-msa" class="section level3" number="18.8.2">
<h3><span class="header-section-number">18.8.2</span> Viewing an MSA</h3>
<p>Once we build an MSA we need to visualize it. There are several ways to do this, and it can be a bit tricky because gene and proteins are long and most easily viewed left to right. Often we’ll identify a subset of bases to focus on, such as a sequence motif or domain.</p>
<div id="viewing-an-msa-in-r" class="section level4" number="18.8.2.1">
<h4><span class="header-section-number">18.8.2.1</span> Viewing an MSA in R</h4>
<p>msa produces a species MSA objects</p>
<div class="sourceCode" id="cb233"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb233-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb233-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(shrooms_align)</span></code></pre></div>
<pre><code>## [1] &quot;MsaAAMultipleAlignment&quot;
## attr(,&quot;package&quot;)
## [1] &quot;msa&quot;</code></pre>
<div class="sourceCode" id="cb235"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb235-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb235-1" aria-hidden="true" tabindex="-1"></a><span class="fu">is</span>(shrooms_align)</span></code></pre></div>
<pre><code>## [1] &quot;MsaAAMultipleAlignment&quot; &quot;AAMultipleAlignment&quot;    &quot;MsaMetaData&quot;           
## [4] &quot;MultipleAlignment&quot;</code></pre>
<p>We can look at the direct output from <code>msa()</code>, but its not very helpful - its just a glimpse of part of the alignment. The “…” in the middle just means “a lot of other stuff in the middle.”</p>
<div class="sourceCode" id="cb237"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb237-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb237-1" aria-hidden="true" tabindex="-1"></a>shrooms_align</span></code></pre></div>
<pre><code>## CLUSTAL 2.1  
## 
## Call:
##    msa(shrooms_vector_ss, method = &quot;ClustalW&quot;)
## 
## MsaAAMultipleAlignment with 14 rows and 2252 columns
##      aln                                                   names
##  [1] -------------------------...------------------------- NP_065768
##  [2] -------------------------...------------------------- AAK95579
##  [3] -------------------------...SVFGGVFPTLTSPL----------- AAF13269
##  [4] -------------------------...SVFGGVFPTLTSPL----------- AAF13270
##  [5] -------------------------...CTFSGIFPTLTSPL----------- NP_065910
##  [6] -------------------------...NKS--LPPPLTSSL----------- ABD59319
##  [7] -------------------------...------------------------- CAA58534
##  [8] -------------------------...------------------------- ABD19518
##  [9] -------------------------...LT----------------------- NP_597713
## [10] -------------------------...------------------------- CAA78718
## [11] -------------------------...------------------------- EAA12598
## [12] -------------------------...------------------------- ABA81834
## [13] MTELQPSPPGYRVQDEAPGPPSCPP...------------------------- XP_392427
## [14] -------------------------...AATSSSSNGIGGPEQLNSNATSSYC XP_783573
##  Con -------------------------...------------------------- Consensus</code></pre>
<p>A function called <code>print_msa()</code> (Coghlan 2011) which I’ve put into<code>combio4all</code> can give us more informative output by printing out the actual alignment into the R console.</p>
<p>To use <code>print_msa()</code> We need to make a few minor tweaks though first to the output of the msa() function. These are behind the scenes changes so don’t worry about the details right now. We’ll change the name to <code>shrooms_align_seqinr</code> to indicate that one of our changes is putting this into a format defined by the bioinformatics package <code>seqinr</code>.</p>
<p>First, we change <strong>class</strong> of the variable to let our functions know exactly what we’re working with.</p>
<p>The output of the class() function can sometimes be a bit complicated; in this case its telling us that the “class” of the shrooms_align object is “MsaAAMultipleAlignment”, which is a special purpose type of R object created by the msa() function (“Msa…”) for amino acids (…“AA”…) that is a multiple sequence alignment (“…MultipleAlignment”)</p>
<div class="sourceCode" id="cb239"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb239-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb239-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(shrooms_align) </span></code></pre></div>
<pre><code>## [1] &quot;MsaAAMultipleAlignment&quot;
## attr(,&quot;package&quot;)
## [1] &quot;msa&quot;</code></pre>
<p>To make shrooms_align play nice with our other functions it has to just be of the class “AAMultipleAlignment”. (Again, this is annoying, and took my a while to figure out when I was creating this workflow).</p>
<p>You rarely have to change the class of an R object; the usual use of the class() function is to just get an idea of what an object is. You can also use the class() function to change the class of an object.</p>
<div class="sourceCode" id="cb241"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb241-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb241-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(shrooms_align) <span class="ot">&lt;-</span> <span class="st">&quot;AAMultipleAlignment&quot;</span></span></code></pre></div>
<p>Now we need to use a function from msa called msaConvert() to make <em>another</em> tweak to the object to make it work with functions from the seqinr package. We’ll change the name of our msa object from shrooms_align to shrooms_align_seqinr to reflect this change. (Another annoying step that took me a while to figure out when I first did this.)</p>
<div class="sourceCode" id="cb242"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb242-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb242-1" aria-hidden="true" tabindex="-1"></a>shrooms_align_seqinr <span class="ot">&lt;-</span> <span class="fu">msaConvert</span>(shrooms_align, </span>
<span id="cb242-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb242-2" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">type =</span> <span class="st">&quot;seqinr::alignment&quot;</span>)</span></code></pre></div>
<p>I won’t display the raw output from <code>shrooms_align_seqinr</code> because its very long; we have 14 shroom genes, and shroom happens to be a rather long gene.
<!-- TODO for Dr. B: set up for displaying only part of the alignment, subsetting it etc. --></p>
<p>Now that I’ve done the necessary tweaks let me display the msa. This will only be really useful if I have a big monitor.</p>
<div class="sourceCode" id="cb243"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb243-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb243-1" aria-hidden="true" tabindex="-1"></a>compbio4all<span class="sc">::</span><span class="fu">print_msa</span>(<span class="at">alignment =</span> shrooms_align_seqinr, </span>
<span id="cb243-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb243-2" aria-hidden="true" tabindex="-1"></a>          <span class="at">chunksize =</span> <span class="dv">60</span>)</span></code></pre></div>
</div>
<div id="displaying-an-msa-as-an-r-plot" class="section level4" number="18.8.2.2">
<h4><span class="header-section-number">18.8.2.2</span> Displaying an MSA as an R plot</h4>
<p>Printing and msa to the R plot window can be useful to make nicer looking figures of parts of an MSA. I’m going to just show about 100 amino acids near the end of the alignment, where there is the most overlap across all of the sequences. This is set with the <code>start = ...</code> and <code>end = ...</code> arguments.</p>
<p>Note that we’re using the <code>shrooms_align</code> object again, but with the class reassigned.</p>
<div class="sourceCode" id="cb244"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb244-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb244-1" aria-hidden="true" tabindex="-1"></a><span class="co"># key step - must have class set properly</span></span>
<span id="cb244-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb244-2" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(shrooms_align) <span class="ot">&lt;-</span> <span class="st">&quot;AAMultipleAlignment&quot;</span></span>
<span id="cb244-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb244-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb244-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb244-4" aria-hidden="true" tabindex="-1"></a><span class="co"># run ggmsa</span></span>
<span id="cb244-5"><a href="worked-example-building-a-phylogeny-in-r.html#cb244-5" aria-hidden="true" tabindex="-1"></a>ggmsa<span class="sc">::</span><span class="fu">ggmsa</span>(shrooms_align,   <span class="co"># shrooms_align, NOT shrooms_align_seqinr</span></span>
<span id="cb244-6"><a href="worked-example-building-a-phylogeny-in-r.html#cb244-6" aria-hidden="true" tabindex="-1"></a>      <span class="at">start =</span> <span class="dv">2000</span>, </span>
<span id="cb244-7"><a href="worked-example-building-a-phylogeny-in-r.html#cb244-7" aria-hidden="true" tabindex="-1"></a>      <span class="at">end =</span> <span class="dv">2100</span>) </span></code></pre></div>
</div>
<div id="saving-an-msa-as-pdf" class="section level4" number="18.8.2.3">
<h4><span class="header-section-number">18.8.2.3</span> Saving an MSA as PDF</h4>
<p>We can take a look at the alignment in PDF format if we want. I this case I’m going to just show about 100 amino acids near the end of the alignment, where there is the most overlap across all of the sequences. This is set with the <code>y = c(...)</code> argument.</p>
<p>In order for this to work you need to have a program called LaTex installed on your computer. LaTex can occasionally be tricky to install so you can skip this step if necessary.</p>
<p>If you want to try to install LaTex, you can run this code to see if it works
for you</p>
<div class="sourceCode" id="cb245"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb245-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb245-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;tinytex&quot;</span>)</span>
<span id="cb245-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb245-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install_tinytex</span>()</span></code></pre></div>
<p>If you have LaTex working on your computer you can then run this code. (If this code doesn’t work, you can comment it out)</p>
<div class="sourceCode" id="cb246"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb246-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb246-1" aria-hidden="true" tabindex="-1"></a><span class="fu">msaPrettyPrint</span>(shrooms_align,             <span class="co"># alignment</span></span>
<span id="cb246-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb246-2" aria-hidden="true" tabindex="-1"></a>               <span class="at">file =</span> <span class="st">&quot;shroom_msa.pdf&quot;</span>,   <span class="co"># file name</span></span>
<span id="cb246-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb246-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">y=</span><span class="fu">c</span>(<span class="dv">2000</span>, <span class="dv">2100</span>),           <span class="co"># range</span></span>
<span id="cb246-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb246-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">askForOverwrite=</span><span class="cn">FALSE</span>)</span></code></pre></div>
<p>You can see where R is saving things by running <code>getwd()</code></p>
<div class="sourceCode" id="cb247"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb247-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb247-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getwd</span>()</span></code></pre></div>
<pre><code>## [1] &quot;/Users/nlb24/OneDrive - University of Pittsburgh/0-books/lbrb-bk/lbrb&quot;</code></pre>
<p>On a Mac you can usually find the file by searching in Finder for the file name, which I set to be “shroom_msa.pdf” using the <code>file = ...</code> argument above.</p>
</div>
</div>
</div>
<div id="a-subset-of-sequences" class="section level2" number="18.9">
<h2><span class="header-section-number">18.9</span> A subset of sequences</h2>
<p>To make things easier we’ll move forward with just a subset of sequences:</p>
<ul>
<li>XP_392427: amShroom (bee shroom)</li>
<li>EAA12598: agShroom (mosquito shroom)</li>
<li>ABA81834: dmShroom (<em>Drosophila</em> shroom)</li>
<li>XP_783573: spShroom (sea urchin shroom)</li>
<li>CAA78718: xShroom1 (frog shroom)</li>
</ul>
<p>Our main working object shrooms_vector_ss has the names of our genes listed</p>
<div class="sourceCode" id="cb249"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb249-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb249-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(shrooms_vector_ss)</span></code></pre></div>
<pre><code>##  [1] &quot;CAA78718&quot;  &quot;NP_597713&quot; &quot;CAA58534&quot;  &quot;ABD19518&quot;  &quot;AAF13269&quot;  &quot;AAF13270&quot; 
##  [7] &quot;NP_065910&quot; &quot;ABD59319&quot;  &quot;NP_065768&quot; &quot;AAK95579&quot;  &quot;ABA81834&quot;  &quot;EAA12598&quot; 
## [13] &quot;XP_392427&quot; &quot;XP_783573&quot;</code></pre>
<p>We can select the ones we want to focus on be first making a vector of the names</p>
<div class="sourceCode" id="cb251"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb251-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb251-1" aria-hidden="true" tabindex="-1"></a>names.focal <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;XP_392427&quot;</span>,<span class="st">&quot;EAA12598&quot;</span>,<span class="st">&quot;ABA81834&quot;</span>,<span class="st">&quot;XP_783573&quot;</span>,<span class="st">&quot;CAA78718&quot;</span>)</span></code></pre></div>
<p>We can use this vector and bracket notation to select the what we want from shrooms_vector_ss:</p>
<div class="sourceCode" id="cb252"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb252-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb252-1" aria-hidden="true" tabindex="-1"></a>shrooms_vector_ss[names.focal]</span></code></pre></div>
<pre><code>## AAStringSet object of length 5:
##     width seq                                               names               
## [1]  2126 MTELQPSPPGYRVQDEAPGPPSC...GREIQDKVKLGEEQLAALREAID XP_392427
## [2]   674 IPFSSSPKNRSNSKASYLPRQPR...ADKIKLGEEQLAALKDTLVQSEC EAA12598
## [3]  1576 MKMRNHKENGNGSEMGESTKSLA...AVRIKGSEEQLSSLSDALVQSDC ABA81834
## [4]  1661 MMKDAMYPTTTSTTSSSVNPLPK...TSSSSNGIGGPEQLNSNATSSYC XP_783573
## [5]  1420 MSAFGNTIERWNIKSTGVIAGLG...KNLEEKIKVYEEQFESIHNSLPP CAA78718</code></pre>
<p>Let’s assign the subset of sequences to a new object called shrooms_vector_ss_subset.</p>
<div class="sourceCode" id="cb254"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb254-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb254-1" aria-hidden="true" tabindex="-1"></a>shrooms_vector_ss_subset <span class="ot">&lt;-</span> shrooms_vector_ss[names.focal]</span></code></pre></div>
<p>Let’s make another MSA with just this subset. If msa isn’t working for you you can comment this out.</p>
<div class="sourceCode" id="cb255"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb255-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb255-1" aria-hidden="true" tabindex="-1"></a>shrooms_align_subset <span class="ot">&lt;-</span> <span class="fu">msa</span>(shrooms_vector_ss_subset,</span>
<span id="cb255-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb255-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">method =</span> <span class="st">&quot;ClustalW&quot;</span>)</span></code></pre></div>
<pre><code>## use default substitution matrix</code></pre>
<p>To view it using ggmsa we need to do those annoying conversions again.</p>
<div class="sourceCode" id="cb257"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb257-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb257-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(shrooms_align_subset) <span class="ot">&lt;-</span> <span class="st">&quot;AAMultipleAlignment&quot;</span></span>
<span id="cb257-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb257-2" aria-hidden="true" tabindex="-1"></a>shrooms_align_subset_seqinr <span class="ot">&lt;-</span> <span class="fu">msaConvert</span>(shrooms_align_subset, <span class="at">type =</span> <span class="st">&quot;seqinr::alignment&quot;</span>)</span></code></pre></div>
<p>THen we can plot it</p>
<div class="sourceCode" id="cb258"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb258-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb258-1" aria-hidden="true" tabindex="-1"></a>ggmsa<span class="sc">::</span><span class="fu">ggmsa</span>(shrooms_align_subset,   <span class="co"># shrooms_align, NOT shrooms_align_seqinr</span></span>
<span id="cb258-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb258-2" aria-hidden="true" tabindex="-1"></a>      <span class="at">start =</span> <span class="dv">2030</span>, </span>
<span id="cb258-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb258-3" aria-hidden="true" tabindex="-1"></a>      <span class="at">end =</span> <span class="dv">2100</span>) </span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-182-1.png" width="672" /></p>
<p>We can save our new smaller MSA like this.</p>
<div class="sourceCode" id="cb259"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb259-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb259-1" aria-hidden="true" tabindex="-1"></a><span class="fu">msaPrettyPrint</span>(shrooms_align_subset,             <span class="co"># alignment</span></span>
<span id="cb259-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb259-2" aria-hidden="true" tabindex="-1"></a>               <span class="at">file =</span> <span class="st">&quot;shroom_msa_subset.pdf&quot;</span>,   <span class="co"># file name</span></span>
<span id="cb259-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb259-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">y=</span><span class="fu">c</span>(<span class="dv">2030</span>, <span class="dv">2100</span>),           <span class="co"># range</span></span>
<span id="cb259-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb259-4" aria-hidden="true" tabindex="-1"></a>               <span class="at">askForOverwrite=</span><span class="cn">FALSE</span>)</span></code></pre></div>
</div>
<div id="genetic-distances-of-sequence-in-subset" class="section level2" number="18.10">
<h2><span class="header-section-number">18.10</span> Genetic distances of sequence in subset</h2>
<p>While an MSA is a good way to examine a sequence its hard to assess all of the information visually. A phylogenetic tree allows you to summarize patterns in an MSA. The fastest way to make phylogenetic trees to is first summarize an MSA using a <strong>genetic distance matrix</strong>. The more amino acids that are identical to each other, the smaller the genetic distance is between them and the less evolution has occurred.</p>
<p>We usually work in terms of <em>difference</em> or <strong>genetic distance</strong> (a.k.a. <strong>evolutionary distance</strong>), though often we also talk in terms of similarity or identity.</p>
<p>Calculating genetic distance from an MSA is done using the <code>seqinr::dist.alignment()</code> function.</p>
<div class="sourceCode" id="cb260"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb260-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb260-1" aria-hidden="true" tabindex="-1"></a>shrooms_subset_dist <span class="ot">&lt;-</span> seqinr<span class="sc">::</span><span class="fu">dist.alignment</span>(shrooms_align_subset_seqinr, </span>
<span id="cb260-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb260-2" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">matrix =</span> <span class="st">&quot;identity&quot;</span>)</span></code></pre></div>
<p>This produces a “dist” class object.</p>
<div class="sourceCode" id="cb261"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb261-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb261-1" aria-hidden="true" tabindex="-1"></a><span class="fu">is</span>(shrooms_subset_dist)</span></code></pre></div>
<pre><code>## [1] &quot;dist&quot;     &quot;oldClass&quot;</code></pre>
<div class="sourceCode" id="cb263"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb263-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb263-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(shrooms_subset_dist)</span></code></pre></div>
<pre><code>## [1] &quot;dist&quot;</code></pre>
<p>If you’ve been having trouble with the MSA software, the data necessary to build the distance matrix directly in R is in this code chunk (you can ignore the details).</p>
<div class="sourceCode" id="cb265"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb265-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb265-1" aria-hidden="true" tabindex="-1"></a>shrooms_subset_dist_alt <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">data =</span> <span class="cn">NA</span>,</span>
<span id="cb265-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb265-2" aria-hidden="true" tabindex="-1"></a>                              <span class="at">nrow =</span> <span class="dv">5</span>, </span>
<span id="cb265-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb265-3" aria-hidden="true" tabindex="-1"></a>                              <span class="at">ncol =</span> <span class="dv">5</span>)</span>
<span id="cb265-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb265-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb265-5"><a href="worked-example-building-a-phylogeny-in-r.html#cb265-5" aria-hidden="true" tabindex="-1"></a>distances <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fl">0.8260049</span>, </span>
<span id="cb265-6"><a href="worked-example-building-a-phylogeny-in-r.html#cb265-6" aria-hidden="true" tabindex="-1"></a>               <span class="fl">0.8478722</span>, <span class="fl">0.9000568</span>, </span>
<span id="cb265-7"><a href="worked-example-building-a-phylogeny-in-r.html#cb265-7" aria-hidden="true" tabindex="-1"></a>               <span class="fl">0.9244596</span>, <span class="fl">0.9435187</span>, <span class="fl">0.9372139</span>, </span>
<span id="cb265-8"><a href="worked-example-building-a-phylogeny-in-r.html#cb265-8" aria-hidden="true" tabindex="-1"></a>               <span class="fl">0.9238779</span>, <span class="fl">0.9370038</span>, <span class="fl">0.9323225</span>,<span class="fl">0.9413209</span>)</span>
<span id="cb265-9"><a href="worked-example-building-a-phylogeny-in-r.html#cb265-9" aria-hidden="true" tabindex="-1"></a>shrooms_subset_dist_alt[<span class="fu">lower.tri</span>(shrooms_subset_dist_alt)] <span class="ot">&lt;-</span> distances</span>
<span id="cb265-10"><a href="worked-example-building-a-phylogeny-in-r.html#cb265-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb265-11"><a href="worked-example-building-a-phylogeny-in-r.html#cb265-11" aria-hidden="true" tabindex="-1"></a>seqnames <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;EAA12598&quot;</span>,<span class="st">&quot;ABA81834&quot;</span>,<span class="st">&quot;XP_392427&quot;</span>, <span class="st">&quot;XP_783573&quot;</span>,<span class="st">&quot;CAA78718&quot;</span>)</span>
<span id="cb265-12"><a href="worked-example-building-a-phylogeny-in-r.html#cb265-12" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(shrooms_subset_dist_alt) <span class="ot">&lt;-</span> seqnames</span>
<span id="cb265-13"><a href="worked-example-building-a-phylogeny-in-r.html#cb265-13" aria-hidden="true" tabindex="-1"></a><span class="fu">row.names</span>(shrooms_subset_dist_alt)  <span class="ot">&lt;-</span> seqnames</span>
<span id="cb265-14"><a href="worked-example-building-a-phylogeny-in-r.html#cb265-14" aria-hidden="true" tabindex="-1"></a>shrooms_subset_dist_alt <span class="ot">&lt;-</span> <span class="fu">as.dist</span>(shrooms_subset_dist_alt)</span>
<span id="cb265-15"><a href="worked-example-building-a-phylogeny-in-r.html#cb265-15" aria-hidden="true" tabindex="-1"></a>shrooms_subset_dist <span class="ot">&lt;-</span> shrooms_subset_dist_alt</span></code></pre></div>
<p>We’ve made a matrix using <code>dist.alignment()</code>; let’s round it off so its easier to look at using the <code>round()</code> function.</p>
<div class="sourceCode" id="cb266"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb266-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb266-1" aria-hidden="true" tabindex="-1"></a>shrooms_subset_dist_rounded <span class="ot">&lt;-</span> <span class="fu">round</span>(shrooms_subset_dist,</span>
<span id="cb266-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb266-2" aria-hidden="true" tabindex="-1"></a>                              <span class="at">digits =</span> <span class="dv">3</span>)</span></code></pre></div>
<p>If we want to look at it we can type</p>
<div class="sourceCode" id="cb267"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb267-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb267-1" aria-hidden="true" tabindex="-1"></a>shrooms_subset_dist_rounded</span></code></pre></div>
<pre><code>##           EAA12598 ABA81834 XP_392427 XP_783573
## ABA81834     0.826                             
## XP_392427    0.848    0.944                    
## XP_783573    0.900    0.937     0.937          
## CAA78718     0.924    0.924     0.932     0.941</code></pre>
<p>Not that we have 5 sequence, but the matrix is 4 x 4. This is because redundant information is dropped, including distances from one sequence to itself. This makes it so that the first column is EAA12598, but the first row is ABA81834.</p>
</div>
<div id="phylognetic-trees-of-subset-sequences-finally" class="section level2" number="18.11">
<h2><span class="header-section-number">18.11</span> Phylognetic trees of subset sequences (finally!)</h2>
<p>We got our sequences, built a multiple sequence alignment, and calculated the genetic distance between sequences. Now we are - finally - ready to build a phylogenetic tree.</p>
<p>First, we let R figure out the structure of the tree. There are <strong>MANY</strong> ways to build phylogenetic trees. We’ll use a common one used for exploring sequences called <strong>neighbor joining</strong> algorithm via the function <code>nj()</code>. Neighbor joining uses genetic distances to cluster sequences into <strong>clades</strong>.</p>
<p>nj() is simple function that takes only a single argument, a distance matrix.</p>
<div class="sourceCode" id="cb269"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb269-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb269-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Note - not using rounded values</span></span>
<span id="cb269-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb269-2" aria-hidden="true" tabindex="-1"></a>tree_subset <span class="ot">&lt;-</span> <span class="fu">nj</span>(shrooms_subset_dist)</span></code></pre></div>
<div id="plotting-phylogenetic-trees" class="section level3" number="18.11.1">
<h3><span class="header-section-number">18.11.1</span> Plotting phylogenetic trees</h3>
<p>Now we’ll make a quick plot of our tree using <code>plot()</code> (and add a little label using an important function called <code>mtext()</code>).</p>
<div class="sourceCode" id="cb270"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb270-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb270-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot tree</span></span>
<span id="cb270-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb270-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot.phylo</span>(tree_subset, <span class="at">main=</span><span class="st">&quot;Phylogenetic Tree&quot;</span>, </span>
<span id="cb270-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb270-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">type =</span> <span class="st">&quot;unrooted&quot;</span>, </span>
<span id="cb270-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb270-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">use.edge.length =</span> F)</span>
<span id="cb270-5"><a href="worked-example-building-a-phylogeny-in-r.html#cb270-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb270-6"><a href="worked-example-building-a-phylogeny-in-r.html#cb270-6" aria-hidden="true" tabindex="-1"></a><span class="co"># add label</span></span>
<span id="cb270-7"><a href="worked-example-building-a-phylogeny-in-r.html#cb270-7" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="at">text =</span> <span class="st">&quot;Shroom family gene tree - UNrooted, no branch lengths&quot;</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-190-1.png" width="672" /></p>
<p>This is an <strong>unrooted tree</strong> with no outgroup defined. For the sake of plotting we’ve also ignored the evolutionary distance between the sequences, so the branch lengths don’t have meaning.</p>
<p>To make a rooted tree we remove <code>type = "unrooted</code>. In the case of neighbor joining, the algorithm tries to figure out the outgroup on its own.</p>
<div class="sourceCode" id="cb271"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb271-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb271-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot tree</span></span>
<span id="cb271-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb271-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot.phylo</span>(tree_subset, <span class="at">main=</span><span class="st">&quot;Phylogenetic Tree&quot;</span>, </span>
<span id="cb271-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb271-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">use.edge.length =</span> F)</span>
<span id="cb271-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb271-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb271-5"><a href="worked-example-building-a-phylogeny-in-r.html#cb271-5" aria-hidden="true" tabindex="-1"></a><span class="co"># add label</span></span>
<span id="cb271-6"><a href="worked-example-building-a-phylogeny-in-r.html#cb271-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="at">text =</span> <span class="st">&quot;Shroom family gene tree - rooted, no branch lenths&quot;</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-191-1.png" width="672" /></p>
<p>We can include information about branch length by setting <code>use.edge.length = ...</code> to <code>T</code>.</p>
<div class="sourceCode" id="cb272"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb272-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb272-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot tree</span></span>
<span id="cb272-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb272-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot.phylo</span>(tree_subset, <span class="at">main=</span><span class="st">&quot;Phylogenetic Tree&quot;</span>, </span>
<span id="cb272-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb272-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">use.edge.length =</span> T)</span>
<span id="cb272-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb272-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb272-5"><a href="worked-example-building-a-phylogeny-in-r.html#cb272-5" aria-hidden="true" tabindex="-1"></a><span class="co"># add label</span></span>
<span id="cb272-6"><a href="worked-example-building-a-phylogeny-in-r.html#cb272-6" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="at">text =</span> <span class="st">&quot;Shroom family gene tree - rooted, with branch lenths&quot;</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-192-1.png" width="672" /></p>
<p>Now the length of the branches indicates the evolutionary distance between sequences and correlate to the distances reported in our distance matrix. The branches are all very long, indicating that these genes have been evolving independently for many millions of years.</p>
<p>An important note: the vertical lines on the tree have no meaning, only the horizontal ones.</p>
<p>Because the branch lengths are all so long I find this tree a bit hard to view when its rooted. Let’s make it unrooted again.</p>
<div class="sourceCode" id="cb273"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb273-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb273-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot tree</span></span>
<span id="cb273-2"><a href="worked-example-building-a-phylogeny-in-r.html#cb273-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot.phylo</span>(tree_subset, <span class="at">main=</span><span class="st">&quot;Phylogenetic Tree&quot;</span>, </span>
<span id="cb273-3"><a href="worked-example-building-a-phylogeny-in-r.html#cb273-3" aria-hidden="true" tabindex="-1"></a>           <span class="at">type =</span> <span class="st">&quot;unrooted&quot;</span>,</span>
<span id="cb273-4"><a href="worked-example-building-a-phylogeny-in-r.html#cb273-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">use.edge.length =</span> T)</span>
<span id="cb273-5"><a href="worked-example-building-a-phylogeny-in-r.html#cb273-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb273-6"><a href="worked-example-building-a-phylogeny-in-r.html#cb273-6" aria-hidden="true" tabindex="-1"></a><span class="co"># add label</span></span>
<span id="cb273-7"><a href="worked-example-building-a-phylogeny-in-r.html#cb273-7" aria-hidden="true" tabindex="-1"></a><span class="fu">mtext</span>(<span class="at">text =</span> <span class="st">&quot;Shroom family gene tree - rooted, with branch lenths&quot;</span>)</span></code></pre></div>
<p><img src="lbrb_files/figure-html/unnamed-chunk-193-1.png" width="672" /></p>
<p>Now you can see that the ABA and EAA sequences form a clade, and that the distance between them is somewhat smaller than the distance between other sequences. If we go back to our original distance matrix, we can see that the smallest genetic distance is between ABA and EAA at 0.826.</p>
<div class="sourceCode" id="cb274"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb274-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb274-1" aria-hidden="true" tabindex="-1"></a>shrooms_subset_dist_rounded</span></code></pre></div>
<pre><code>##           EAA12598 ABA81834 XP_392427 XP_783573
## ABA81834     0.826                             
## XP_392427    0.848    0.944                    
## XP_783573    0.900    0.937     0.937          
## CAA78718     0.924    0.924     0.932     0.941</code></pre>
<p>We can confirm that this is the minimum using the min() function.</p>
<div class="sourceCode" id="cb276"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb276-1"><a href="worked-example-building-a-phylogeny-in-r.html#cb276-1" aria-hidden="true" tabindex="-1"></a><span class="fu">min</span>(shrooms_subset_dist_rounded)</span></code></pre></div>
<pre><code>## [1] 0.826</code></pre>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="a-complete-bioinformatics-workflow-in-r.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="simple-for-loop-example.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/brouwern/lbrb/edit/master/005-MSA/23-MSA-walkthrough-shroom.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": "https://github.com/brouwern/lbrb/blob/master/005-MSA/23-MSA-walkthrough-shroom.Rmd",
"text": null
},
"download": ["lbrb.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
